<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
<html> <head>
<title>GBrowse Tutorial</title>
<link rel="stylesheet" href="tutorial.css">
</head>

<body>
<h1>Generic Genome Browser: A Tutorial</h1>

<h2>Author: Lincoln Stein, June 16, 2003 (<i>revised 20 Mar 2007</i>)</h2>

<p>

This is a quick tutorial to take you through the main features and
gotchas of GBrowse.  This tutorial assumes that you have successfully
setup Perl, GD, BioPerl and the other GBrowse dependencies.  During
most of the tutorial, we will be using the "in-memory" GBrowse
database (no relational database required!)  Later we will show how to
set up a genome size database using the berkeleydb and MySQL adaptors.
</p>

<p>

We will be working with simulated Volvox genome annotation data.  The
database will be named "volvox" and GBrowse will be invoked with this
URL:

<blockquote
class="example"><pre>http://localhost/cgi-bin/gbrowse/volvox</pre></blockquote>

<p>

These directories contain data files used during the tutorial:

<dl>
  <dt><a href="data_files/">data_files</a>
  <dd>DNA and features files to load into the local database.
      <p>
  <dt><a href="conf_files/">conf_files</a>
  <dd>GBrowse configuration files for you to take and modify.
</dl>

<h2><a name="basics">1. The Basics</a></h2>

<p>

We will be using a file-based database which allows GBrowse to run
directly off text files.  To prepare this database for use, find the
GBrowse databases directory which was created in your Apache web
server directory at the time of installation.  It should be located at
<b>/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases</b>, but check to make sure.

<p>

Similarly, check that you can find the gbrowse.conf configuration
directory.  It should be located at <b>/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/conf/gbrowse.conf</b> and
contain the configuration file "yeast_chr1.conf."

<p>

Now you will change the permissions of the database and configuration
directories so that you can write to them without root privileges.
This is only an issue on Unix systems, and Windows users can safely
ignore this step.

<blockquote class="example"><pre>
% <b>su</b>
Password: <b>*********</b>
# <b>chown my_user_name /tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases</b>
# <b>chown my_user_name /tmp/update-vendor-drop-jcQwDOU/FAKEROOT/conf/gbrowse.conf</b>
# <b>exit</b>
%
</pre></blockquote>

<p>

(Be sure to replace "my_user_name" with your login name!)

<p>

Now look around inside the databases directory.  There should be a
single subdirectory named "yeast_chr1." The yeast subdirectory is
where the example yeast chromosome 1 data set is stored.

<p>

You will create an empty volvox subdirectory, and make it world
writable.  On Unix systems:

<blockquote class="example"><pre>
% <b>cd /tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases</b>
% <b>mkdir volvox</b>
% <b>chmod go+rwx volvox</b>
</pre></blockquote>

<p>

<blockquote> <i>NOTE: The "%" sign in these examples is the
command-line prompt.  On Windows systems, the command-line prompt is
something like C:\Program Files\Apache
Group\Apache2\htdocs\gbrowse\databases&gt;.  Unix systems are more
variable, but the prompt usually ends with a "%" or a "#".  In all the
examples in this tutorial, what you type is rendered in
<b>boldface</b>, while prompts and command-line results are shown in
medium typeface.</i> </blockquote>

<p>

On Windows systems, use the file manager ("Explorer") to create a new
folder named "volvox."  If you are using Windows NT, 2000 or XP, right
click on the new folder and grant write privileges to all.

<p>

You'll now put the first of several data files into the volvox
database directory.  In the <a href="data_files/">data_files</a>
subdirectory of this tutorial you will find the file <a
href="data_files/volvox1.gff">volvox1.gff</a>.  Copy this into the
volvox database directory.  On Unix systems:

<blockquote class="example">
<pre>
% <b>cd /tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse</b>
% <b>cp tutorial/data_files/volvox1.gff databases/volvox</b>
</pre>
</blockquote>

<p>

On Windows systems, use Explorer to copy the file into the volvox
database directory.

<p>

Now we will need a GBrowse config file to tell GBrowse how to render
this data set.  In the subdirectory <a
href="conf_files">conf_files</a>, you will find a sample configuration
file named <a href="conf_files/volvox.conf">volvox.conf</a>.  Copy
this into your GBrowse configuration directory (/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/conf/gbrowse.conf).

<p>

<p>

You should now be able to view the data set.  Point your web browser
at <a
href="/cgi-bin/gbrowse/volvox">http://localhost/cgi-bin/gbrowse/volvox</a>
and type in "ctgA" in the search box.  The result is shown in Figure
1.

<p>
<blockquote>
<img src="figures/basics1.gif"><br>
<i>Figure 1: volvox1.gff data with volvox.conf config file.</i>
</blockquote>

<h3>If You are Having Problems...</h3>

<p>

If for some reason you get a blank page or an "Internal server error,"
there are a couple of things to check.  First, open the file
volvox.conf with a text editor ("Notepad" on Windows systems, emacs,
pico or vi on Unix systems) and confirm that the path to the volvox
database directory in this section is correct:

<pre>
db_args       = -adaptor memory
		-gff    '/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases/volvox'
</pre>

<p>

If there is a space in "/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs" then you must be certain to put
single quotes around the path as shown in the example above.

<p>

Next check that the volvox1.gff file does exist inside the volvox
database directory and that it is readable by all users on your
system.  Similarly, check that the volvox.conf configuration file is
in the same directory as yeast_chr1.conf, and that it is readable by
all users on your system.

<p>

Microsoft Windows has an unpleasant tendency to add a ".txt" extension
to files without warning.  If something seems to be wrong with the
config or GFF file and you can't figure out what, check that the file
extension hasn't been modified.  To avoid this phenomenon, I suggest
that you select "All File Types" from the popup menu in the File Save
dialog.  You might also want to configure your Folder display to show
known file extensions.

<p>

If you're still having no luck, check the bottom of the Apache server
error log for error messages.  This file is located in various places
depending on how Apache is installed.  Look for the file
<b>error_log</b>, typically located in /usr/local/apache/logs,
C:\Program Files\Apache Group\Apache2\logs, /var/log/www, or
/var/log/httpd.  The error message will usually point you in the right
direction.

<p>

<blockquote>
If you're still having no luck, sorry!  Stop the
tutorial and send an e-mail to GBrowse support at <a
href="mailto:gmod-gbrowse@lists.sourceforge.net">gmod-gbrowse@lists.sourceforge.net</a>
</blockquote>

<hr>

<hr>

<h2><a name="data_file">2. The Data File</a></h2>

<p>


Let's look at the data file we loaded in detail now.  If you open the
<a href="data_files/volvox1.gff">volvox1.gff</a> file in a text
editor, you will see that it contains a series of 15 genome "features"
that look like this:

<blockquote class="example"><pre>
ctgA   example   contig       1      50000   .   .   .   Contig ctgA
ctgA   example   my_feature   1659   1984    .   +   .   My_feature f07
ctgA   example   my_feature   3014   6130    .   +   .   My_feature f06
ctgA   example   my_feature   4715   5968    .   -   .   My_feature f05
ctgA   example	 my_feature   13280  16394   .   +   .   My_feature f08
...
</pre></blockquote>

<p>

Each feature has a "source" of example, a type of "my_feature", and
occupies a short range (roughly 1.5k) on a contig named "ctgA."  In
addition to the features themselves, there is an entry for the contig
itself (type "contig").  This entry is needed to tell GBrowse what the
length of ctgA is.

<p>

The load file uses a semi-standard known as GFF (General Feature
Format).  Each line of the file corresponds to a feature on the
genome, and the nine columns are separated by tabs.

<p>

The 9 columns are as follows:

<ol>
  <li><b>reference sequence</b><br>
      This is the name of the feature that will be used to establish the
      coordinate system for the annotation.  This is usually the name of
      a chromosome, a clone, or a contig.  In our example, the
      reference sequence is "ctgA".  A single GFF file can refer to
      multiple reference sequences.</li><br>
  <li><b>source</b><br>
      The source of the annotation.  This field describes how the
      feature was derived.  In the example, the source is
      "example" for want of a better description.  Many people find
      the source as a way of distinguishing between similar features
      that were derived by different methods, for example, gene
      calls derived from different prediction software.  You can
      leave this column blank by replacing the source with a single
      dot (".").</li><br>
  <li><b>type</b><br>
      This column describes the feature type.  You can choose anything
      you like to describe the feature type, but common names are
      "gene", "repeat", "exon", and "CDS."  A good source of commonly
      recognized names is
      the <a
      href="http://song.cvs.sourceforge.net/song/ontology/sofa.ontology?rev=HEAD&content-type=text/vnd.viewcvs-markup">Sequence
      Ontology Lite</a>, located at http://song.sourceforge.net.  For
      lack of a better name, the features in the volvox example are of
      type "my_feature."</li><br>
  <li><b>start position</b><br>
      The position that the feature starts at, relative to the
      reference sequence.  The first base of the reference sequence
      is position 1.</li><br>
  <li><b>end position</b><br>
      The end of the feature, again relative to the reference
      sequence.  End is always greater than or equal to start.</li><br>
  <li><b>score</b><br>
      For features that have a numeric score, such as sequence
      similarities, this field holds the score.  Score units are
      arbitrary, but most people use the expectation value for
      similarity features.  You can leave it blank by replacing
      the column with a dot.</li><br>
  <li><b>strand</b><br>
      For features that are strand-specific, this field is the
      strand on which the annotation resides.  It is "+" for the forward
      strand, "-" for the reverse strand, or "." for annotations that are
      not stranded.  If you are unsure of whether a feature is
      stranded, it won't hurt to use a "+" here.</li><br>
  <li><b>phase</b><br>
      For CDS features that encode proteins, this field describes
      where the next codon starts.
      The phase is one of the integers 0, 1, or 2, indicating the
      number of bases that should be removed from the beginning of
      this feature in order to reach the first base of the next codon. In other
      words, a phase of "0" indicates that the next codon begins at
      the first base of the region described by the current line, a
      phase of "1" indicates that the next codon begins at the second
      base of this region, and a phase of "2" indicates that the next codon
      begins at the third base of this region. This
      information is used by the "cds" glyph to show how the reading
      frame changes across splice sites.  For all other feature types,
      use a dot here.</li><br>
  <li><b>group</b><br>
      The ninth and last column has multiple purposes.  Its main use
      is to give features a name for searching and display.  It can
      also be used to group related features together by giving them a
      common name.  We'll see later how the exons of a gene can be grouped together
      with this field. <br><br>
      The format of the group field is  "<i>class</i> <i>name</i>",
      where <i>class</i> describes the
      class of the feature and <i>name</i> describes its name.  The
      class and the name are separated by a space <b>not</b> a tab.
      The feature class is a funny concept, because it is almost, but not
      quite, the same as the feature type.  The idea is to distinguish
      among features that might share the same name by giving them a
      distinctive prefix.  For example, the class distinguishes
      "Transcript M1.2" from "Gene M1.2".  Most people find this
      confusing, and a proposed update to the GFF format promises to do away
      with the class entirely.  For now, I suggest that you reuse the type
      field here.  In the examples, I've used an initial capital
      letter in the class field in order to distinguish when I'm
      talking about the class field from when I'm talking about the
      type field.
      <br><br>
      Other good stuff can go into the group field, as we shall see later.</li>
</ol>

<hr>

<h2><a name="basic_conf">3. Defining Tracks</a></h2>

<p>

Now we'll look at the configuration file in more detail.  Using a text
editor, open the volvox.conf file from its location in the
gbrowse.conf configuraton directory.  (If you mess up, you can always
copy a fresh version from <a
href="conf_files/volvox.conf">volvox.conf</a> in the tutorial
directory).

<p>

Ignore all the stuff in the top 90% of the file, and focus on the last
little bit, which starts with the line: <code>### TRACK CONFIGURATION
###</code>:

<blockquote class="example"><pre>
[ExampleFeatures]
feature      = my_feature
glyph        = generic
stranded     = 1
bgcolor      = blue
height       = 10
key          = Example Features
</pre></blockquote>

<p>

This is a "stanza" that describes one of the tracks displayed by
GBrowse.  The track has an internal name of "ExampleFeatures" which
you can use in the URL to turn the track on.  The internal name is
enclosed by square brackets.

<p>

Following the track name are a series of
options that configure the track.  The "feature" option indicates what
feature type(s) to display inside the track.  It's currently set to
display the "my_feature" feature type.  The "glyph" option specifies
the shape of the rendered feature.  The default is "generic", which is
a simple filled box, but there are dozens of glyphs to choose from.
The "stranded" option tells the generic glyph to try to display the
strandedness of the feature -- this is what creates the little arrow
at the end of the box.  "bgcolor" and "height" control the background
color and height of the glyph respectively, and "key" assigns the
track a human-readable label.

<p>

Let's experiment with changing the track definition.  First, let's
change the color of the glyph.  With your text editor, change the
bgcolor option from blue to "orange", save it, and reload the page.
The features should change immediately as shown in Figure 2.

<blockquote>
<img src="figures/basic_conf1.gif"><br>
<i>Figure 2: A Feature of a Different Color</i>
</blockquote>

<p>

Please experiment with other changes!   Try changing the height to 5, the
key to "Skinny features" and the stranded option to 0 (which means
"false").  Just by changing a few options, you can create a very
distinctive track.

<p>

Now let's try changing the glyph.  One of the standard glyphs was
designed to show PCR primer pairs and is called "primers".  Change
"glyph = generic" to "glyph = primers" and reload the page.  Depending
on other changes that you might have made earlier, the result will
look something like Figure 3.

<p>


<blockquote>
<img src="figures/basic_conf2.gif"><br>
<i>Figure 3: Using the primers Glyph</i>
</blockquote>

<p>

We'll see other examples of glyphs later on.  To get a list of the
most popular glyphs and the options that are available for them, see
the file CONFIGURE_HOWTO.txt, located in the docs/ subdirectory of the
GBrowse distribution.  Or for the gory and bleeding edge details, run
the command:

<blockquote class="example"><pre>
 % <b>perldoc Bio::Graphics::Panel</b>
</pre></blockquote>

<p>

This produces copious documentation on the Perl interface to all the
glyphs, including some amazingly obscure ones, from which you should
be able to deduce the GBrowse equivalents.

<hr>

<h2><a name="naming">4. Searching for Named Features</a></h2>

<p>

Now we'll look at the interaction between feature names and GBrowse's
search box.  If you look through the <a
href="data_files/volvox1.gff">volvox1.gff</a> data file, you'll see
that all the example features are named, and that their class is
"My_feature."

<p>

GBrowse has a very flexible search feature.  You can type in the name
of a reference sequence, such as "ctgA", and it will display the
entire thing, or you can type in a range in the format
"ctgA:start..stop".  Try "ctgA:5000..8000" to see this at work.

<p>

In addition, GBrowse can search for features by name.  By default, the
name of the object must be preceded by its class in the format
<i>Class:name</i>.  For example, if we are searching for My_feature
"f07", we could type "My_feature:f07" into the search box.  Try this
now.

<p>

You probably don't want to remember to enter the class of the object
to search for a feature.  Fortunately, it is easy to declare one or
more classes "automatic" and specify the order in which GBrowse will
search for them.  To do this with our example database, open up the
volvox.conf config file, find the option named "automatic classes",
and change it to read:

<blockquote class="example"><pre>
automatic classes = My_feature
</pre></blockquote>

<p>

This tells GBrowse that when the user types in an unqualified feature
name, it should search the My_feature class for a match.  You can now
type "f07" directly into the search field and GBrowse will find and
display it.  If you wish, you may list several (or many!) automatic classes on
this line.  Just separate them by spaces:

<blockquote class="example"><pre>
automatic classes = My_feature Gene Transcript Contig Chromosome
</pre></blockquote>

<p>

For fun, try searching for the following:

<ul>
  <li>f1*
  <li>f07:-5000..5000
  <li>*3
</ul>
<small><sup>Note</sup>These searches may not work as expected
when using the memory adaptor due to a bug in bioperl 1.2.3.  It is fixed in 
bioperl 1.4.</small>

<p>

<hr>

<h2><a name="descriptions">5. Adding Descriptions to a Feature</a></h2>

<p>

By default, GBrowse will display the name of the feature above its
glyph provided that there is sufficient space to do this.  Optionally,
you can also attach some descriptive text to the feature.  This text
will be displayed below the feature, and can also be searched.

<p>

You can place descriptions, notes and other comments into the ninth
column of the GFF load file.  The example file <a
href="data_files/volvox2.gff">volvox2.gff</a> shows how this is done.
An excerpt from the top of the file looks like this:

<blockquote class="example"><pre>
ctgA   example   motif   11911   15561   .   +   .   Motif m11 ; Note "kinase"
ctgA   example   motif   13801   14007   .   -   .   Motif m05 ; Note "helix loop helix"
ctgA   example   motif   14731   17239   .   -   .   Motif m14 ; Note "kinase"
ctgA   example   motif   15396   16159   .   +   .   Motif m03 ; Note "zinc finger"
</pre></blockquote>

<p>

This defines several new features of type "motif".  The ninth column,
in addition to giving each of the motifs names (e.g. "Motif m11"),
adds a "Note" <i>attribute</i> to each feature.  Attributes are
name/value pairs, where the attribute name is a single word, and the
value is a piece of text.  If the value text contains white space
(spaces or tabs), the text must be enclosed by double quotes as shown
above.  It doesn't hurt to enclose the text in quotes even if it
doesn't contain white space, which is what we've done here.  Attribute
pairs are separated from the Class/name pair by a semicolon, as shown
above.

<p>

We will see later how to take advantage of different types of
attributes.  For now, it is only important to know that an attribute
named <i>Note</i> is automatically displayed and made searchable.  To
see this work, add <a href="data_files/volvox2.gff">volvox2.gff</a>
to the volvox database.  You can do this just by copying the file
into <b>/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases/volvox</b> so that the directory
contains both the original volvox1.gff and the new volvox2.gff files.

<p>

To display this newly-loaded data set, open up volvox.conf and add the
following new stanza to the config file:

<blockquote class="example"><pre>
[Motifs]
feature      = motif
glyph        = span
height       = 5
description  = 1
key          = Example motifs
</pre></blockquote>

<p>

This defines a new track whose internal name is "Motifs."  The
corresponding feature type is "motif" and it uses the "span" glyph, a
graphic that displays a horizontal line capped by vertical endpoints.
The height is set to five pixels, and the human-readable key is set to
"Example motifs."  A new option, "description" is a flag that tells
GBrowse to display the Note attribute, if any.  Any non-zero value
means true.

<p>

After updating the configuration file, you will need to reload the
browser page and turn on the "Example motifs" checkbox below the main
image. The result is shown in figure 4.

<blockquote>
<img src="figures/descriptions1.gif"><br>
<i>Figure 4: Showing the Notes attribute</i>
</blockquote>

<p>

A copy of this config file is also available for you to use in
<a href="conf_files/volvox2.conf">volvox2.conf</a>.

<p>

To show that GBrowse will search the notes for keyword matches, try
typing in "kinase."  You will be presented with a list of all the
motifs whose Note attribute contains the word "kinase."

<hr>

<h2><a name="segmented_features">6. Showing multi-segmented features</a></h2>

<p>

Many features are discontinuous.  Examples include spliced
transcripts, and gapped sequence similarity alignments, such as the
alignment of cDNAs to the genome.  GBrowse can deal with such features
easily provided that you take a little care in setting them up.

<p>

The data file <a href="data_files/volvox3.gff">volvox3.gff</a>
contains a simulated data set of a series of gapped nucleotide
alignments.  An excerpt from the file is here:

<blockquote class="example"><pre>
ctgA   example   match 6885   8999   .   -   .   Match seg03
ctgA   example   HSP   6885   7241   .   -   .   Match seg03
ctgA   example   HSP   7410   7737   .   -   .   Match seg03
ctgA   example   HSP   8055   8080   .   -   .   Match seg03
ctgA   example   HSP   8306   8999   .   -   .   Match seg03
ctgA   example   match 5233   9825   .   -   .   Match seg04
ctgA   example   HSP   5233   5302   .   -   .   Match seg04
ctgA   example   HSP   5800   6101   .   -   .   Match seg04
ctgA   example   HSP   6442   6854   .   -   .   Match seg04
ctgA   example   HSP   7106   7211   .   -   .   Match seg04
ctgA   example   HSP   7695   8177   .   -   .   Match seg04
ctgA   example   HSP   8545   8783   .   -   .   Match seg04
ctgA   example   HSP   8869   8935   .   -   .   Match seg04
ctgA   example   HSP   9404   9825   .   -   .   Match seg04
</pre></blockquote>

<p>

Each segmented feature is represented by several lines in the GFF file
that share the same feature name.  Each feature has a single GFF line
of type "match" whose start and end coordinates correspond to the full
length of the alignment.  Following this are one or more lines of type
"HSP" with start and end coordinates indicating a section of the
match.  You will recognize this terminology from the standard BLAST
report.

<p>

For example "Match seg03" starts at position 6885 and ends at 8999.
It has four subsegments, one from 6885..7241, another from 7410..7737,
and so forth.

<p>

The types "match" and "HSP" are not arbitrary, but are needed to tell
the GBrowse database what the relationship between the full-length
feature and its subparts are.  The specific type names expected are
mediated by a series of "aggregators" -- code modules that are loaded
when GBrowse starts up.  We will see later in this section how to
manipulate these aggregators and to define custom ones.

<p>

Add <a href="data_files/volvox3.gff">volvox3.gff</a> into the volvox
database by copying it into the volvox database directory.  Then edit
volvox.conf to add the following track definition:

<blockquote class="example"><pre>
[Alignments]
feature      = match
glyph        = segments
key          = Example alignments
</pre></blockquote>

<p>

This is declaring a new track named "Alignments" which displays
features of type "match" using a glyph named "segments".  The segments
glyph is specialized for displaying objects that have multiple
similar subparts.

<p>

Save the modified config file and reload the page in the browser.
Disappointingly, you'll see something like Figure 5.  Instead of
showing multi-segmented features, the track shows a single solid box
that spans the entire length of the feature.

<blockquote>
<img src="figures/segmented_features1.gif"><br>
<i>Figure 5: Without activating an aggregator, multisegmented features
do not displayed properly</i>
</blockquote>

<p>

To make multipart features display correctly, you must activate or
define an appropriate aggregator.  This is very easy for the
similarity/match relationship, because there's already a predefined
aggregator named "match."  Reopen the volvox.conf configuration file,
and find the option line near the top of the file that reads
"aggregators = ".  Change this to read as follows:

<blockquote class="example"><pre>
aggregators = match
</pre></blockquote>

<p>

This is telling GBrowse to turn on the "match" aggregator.  Now
reload.  You should see a much-improved image similar to Figure 6.

<blockquote>
<img src="figures/segmented_features2.gif">
<i>Figure 6: Turning on the "match" aggregator allows GBrowse to
recognize that the match feature has subparts</i>
</blockquote>

<hr>

<h2><a name="aggregators">7. Using Aggregators</a></h2>

There are several predefined aggregators, each of which expects a
particular combinations of feature type names. The table below
summarizes the most useful ones:

<table border="1">
  <tr>
    <th>Aggregator name</th>
    <th>Main type</th>
    <th>Subtype(s)</th>
    <th>Purpose</th>
  </tr>
  <tr>
    <th>alignment</th>
    <td>(none)</td>
    <td>similarity</td>
    <td>nucleotide and protein alignments where the full extent of the
    match is unknown</td>
  </tr>
  <tr>
    <th>coding</th>
    <td>mRNA</td>
    <td>CDS</td>
    <td>Used in concert with the "cds" glyph to display the reading
	frame used by the coding portion of each exon.</td>
  </tr>
  <tr>
    <th>clone</th>
    <td>(none)</td>
    <td>Clone_left_end Clone_right_end</td>
    <td>Used for cases in which clone ends have been mapped to the
	genome, but one of the ends may be missing.</td>
  </tr>
  <tr>
    <th>match</th>
    <td>match</td>
    <td>similarity, HSP</td>
    <td>nucleotide and protein alignments</td>
  </tr>
  <tr>
    <th>processed_transcript</th>
    <td>mRNA</td>
    <td>CDS, UTR, 5'-UTR, 3'-UTR transcription_start_site, polyA_site</td>
    <td>the canonical spliced gene</td>
  </tr>
  <tr>
    <th>transcript</th>
    <td>transcript</td>
    <td>exon TSS PolyA</td>
    <td>a spliced transcript that expects exon features</td>
  </tr>
</table>

<p>

To use any of these aggregators, follow this recipe:

<ol>
  <li>Give your features and their subparts the specific type names
      expected by the aggregators.
  <li>Add the aggregator to the list of aggregators in the config
      file, e.g. <code>aggregators = match processed_transcript
      clone</code>.
  <li>In the appropriate track definition, use the aggregator's name
      as the argument for <code>feature</code>.  For example
      <code>feature=processed_transcript</code>.
</ol>

<hr>

<h2><a name="canonical_gene">8. Showing Protein Coding Genes</a></h2>

<p>

GBrowse can display protein-coding genes in various shapes and styles.
The easiest way to set this up is to use the "processed_transcript"
aggregator and its companion glyph also called "processed_transcript."
Take a look at the file <a
href="data_files/volvox4.gff">volvox4.gff</a>, which defines a gene
named EDEN, and its three spliced forms named EDEN.1, EDEN.2 and
EDEN.3.  Here is the contents of the file:

<blockquote class="example"><pre>
ctgA    example gene    1050    9000    .       +       .       Gene EDEN ; Note "protein kinase"
                                          
ctgA    example mRNA    1050    9000    .       +       .       mRNA EDEN.1 ; Gene EDEN
ctgA    example 5'-UTR  1050    1200    .       +       .       mRNA EDEN.1
ctgA    example CDS     1201    1500    .       +       0       mRNA EDEN.1
ctgA    example CDS     3000    3902    .       +       0       mRNA EDEN.1
ctgA    example CDS     5000    5500    .       +       0       mRNA EDEN.1
ctgA    example CDS     7000    7608    .       +       0       mRNA EDEN.1
ctgA    example 3'-UTR  7609    9000    .       +       .       mRNA EDEN.1
                                                           
ctgA    example mRNA    1050    9000    .       +       .       mRNA EDEN.2 ; Gene EDEN
ctgA    example 5'-UTR  1050    1200    .       +       .       mRNA EDEN.2
ctgA    example CDS     1201    1500    .       +       0       mRNA EDEN.2
ctgA    example CDS     5000    5500    .       +       0       mRNA EDEN.2
ctgA    example CDS     7000    7608    .       +       0       mRNA EDEN.2
ctgA    example 3'-UTR  7609    9000    .       +       .       mRNA EDEN.2
                                            
ctgA    example mRNA    1300    9000    .       +       .       mRNA EDEN.3 ; Gene EDEN
ctgA    example 5'-UTR  1300    1500    .       +       .       mRNA EDEN.3
ctgA    example 5'-UTR  3000    3300    .       +       .       mRNA EDEN.3
ctgA    example CDS     3301    3902    .       +       0       mRNA EDEN.3
ctgA    example CDS     5000    5500    .       +       1       mRNA EDEN.3
ctgA    example CDS     7000    7600    .       +       1       mRNA EDEN.3
ctgA    example 3'-UTR  7601    9000    .       +       .       mRNA EDEN.3
</pre></blockquote>

<p>

The first line of the file defines the gene as a whole, starting at
position 1050 of ctgA and extending to position 9000.  Following this,
there are three sets of lines that define the structure of the spliced
forms EDEN.1, EDEN.2, and EDEN.3.  By convention, the whole transcript is represented
as type "mRNA".  It has subparts named "5'-UTR", CDS, and "3'-UTR",
where the UTR features are the 5' and 3' untranslated regions,
respectively, and CDS is the coding region.  Note how the CDS is split
by splicing among multiple discontinuous locations on the reference
sequence.  The UTRs can be split in this way too.

<p>

Each mRNA and its subparts are grouped together under a common name in
the ninth column ("mRNA EDEN.1", "mRNA EDEN.2", and so forth).  In
addition, each mRNA has a Gene <i>attribute</i> that ties it to the
EDEN gene itself ("Gene EDEN").  Although this isn't required for the
display, doing this will identify the various alternative transcripts
as belonging to the same gene should you wish to use the GBrowse
database for data mining.  It will also show the user what gene the
transcript belongs to when he or she clicks on it for details.

<p>

<blockquote class="example"><i>HINT: If you prefer not to distinguish between 5' and
3' UTRs, you can simply use "UTR" as the type.  If you don't know
where the UTRs are, just leave them blank.  If you'd rather think in
terms of exons and introns, then check out the "transcript" aggregator
and its corresponding "transcript" glyph.
</i>
</blockquote>

<p>

Go ahead and add volvox4.gff to the database.  Then make the following
changes to volvox.conf:

<ol>
  <li>Change the aggregators line to read as follows:
      <blockquote class="example"><pre>
aggregators = match
	processed_transcript
      </pre></blockquote>
  <li>Add the following new stanza to the bottom of the file:
      <blockquote class="example"><pre>
[Transcripts]
feature      = processed_transcript
glyph        = processed_transcript
bgcolor      = peachpuff
description  = 1
key          = Protein-coding genes
      </pre></blockquote>
</ol>

<p>

The updated aggregators option loads the processed_transcript
aggregator, which knows how to put CDS and UTR features together to
form a spliced transcript.  The new Transcripts track associates
aggregated processed_transcript features with the like-named glyph,
sets its background color to peachpuff (yes, there really is a color
by this name!), turns on the description lines, and sets the human
readable track name to "Protein-coding genes."  <p>

<blockquote class="example"><pre>
<i>The aggregators option demonstrates that GBrowse config file
options can continue across multiple lines
provided that each additional line is indented.</i>
</blockquote>

<p>

Upon reloading the page, turning on the new "Protein-coding genes"
track, and viewing the region around 1..10K, you'll see this:

<blockquote>
<img src="figures/canonical_gene1.gif"><br>
<i>Figure 7: The canonical processed_transcript glyph</i>
</blockquote>

<p>

This image is nice, but we can make it even better.  One problem is
that the gene description (the Note in the EDEN GFF line) isn't being
displayed, because the description is attached to the gene and not to
the individual mRNAs.  To fix this, we simply tell GBrowse to display
features of type "gene" as well as those of type
"processed_transcript".  Modify volvox.conf so the last stanza looks
like this:

<blockquote class="example"><pre>
[Transcripts]
feature      = processed_transcript gene
glyph        = processed_transcript
bgcolor      = peachpuff
description  = 1
key          = Protein-coding genes
</pre></blockquote>

<p>

The only change is that there are now <b>two</b> types listed for the
feature option, "processed_transcript" and "gene."  This is telling
GBrowse to place both feature types in the same track.  If we reload
the page, it now looks like this:

<blockquote>
<img src="figures/canonical_gene2.gif">
<i>Figure 8: Showing the gene as well as its transcripts</i>
</blockquote>

<p>

The processed_transcript glyph has a number of options that you can
use to customize its appearance:
<p>
<table border="1">
  <tr>
    <th>Option Name</th>
    <th>Possible values</th>
    <th>Description</th>
  </tr>
  <tr>
    <th>thin_utr</th>
    <td>0 (false), 1 (true)</td>
    <td>If true, makes UTRs half-height.</td>
  </tr>
  <tr>
    <th>utr_color</th>
    <td>a color name ("gray" by default)</td>
    <td>Changes the UTR color.</td>
  </tr>
    <tr>
    <th>decorate_introns</th>
    <td>0 (false), 1 (true)</td>
    <td>If true, puts little arrowheads on the introns to indicate
	direction of transcription.</td>
  </tr>
</table>

<p>

Using these options, we can make the track look like the UCSC Genome
Browser (Figure 9).

<blockquote class="example"><pre>
[Transcripts]
feature      = processed_transcript gene
glyph        = processed_transcript
height       = 8
bgcolor      = black
utr_color    = black
thin_utr     = 1
decorate_introns = 1
description  = 1
key          = Protein-coding genes
</pre></blockquote>

<p>

<blockquote>
<img src="figures/canonical_gene3.gif"><br>
<i>Figure 9: A UCSC Genome Browser lookalike</i>
</blockquote>


<hr>

<h2><a name="cds">9. Showing the Reading Frame</a></h2>

<p>

Continuing with the example from the last section, the third exon of
EDEN.1 is shared with EDEN.3.  But is the reading frame preserved?
The "coding" aggregator used in concert with the "cds" glyph creates a
display that will visualize each CDS's reading frame.

<p>

To see this work, add the predefined "coding" aggregator to the list
of aggregators:

<blockquote class="example"><pre>
aggregators = match
	      processed_transcript
	      coding
</pre></blockquote>

<p>

The "coding" aggregator is similar to processed_transcript, except
that it only pays attention to the CDS parts of the transcript.  It
was designed to work hand-in-hand with the "cds" glyph.  (For
historical reasons, the glyph is called "cds" rather than "coding.")

<p>

Now add the following short stanza to the bottom of the configuration
file:

<blockquote class="example"><pre>
[CDS]
feature      = coding
glyph        = cds
key          = Frame usage
</pre></blockquote>

<p>

When you reload the page and turn this track on, you'll see a "musical
staff" representation of the frame usage (Figure 10).  From this we
can see that the alternative splicing in fact changes the reading
frame of the second exon.

<blockquote>
<img src="figures/cds1.gif"><br>
<i>Figure 10:  The "cds" glyph shows the reading frame using a musical
staff notation</i>
</blockquote>

<hr>

<h2><a name="custom_aggregators">10. Creating Custom Aggregators</a></h2>
<p>

If none of the predefined aggregators meets your needs, it is simple
to define a custom one of your own.  For example, say you wanted to
display a feature of type "BAC", whose subparts are of type
"left_end_read" and "right_end_read" (possibly corresponding to a BAC
clone mapping experiment).  Here is a GFF representation of this:

<blockquote class="example"><pre>
ctgA   example   BAC              1000   20000   .   .   .   BAC b101.2 ; Note "Fingerprinted BAC with end reads"
ctgA   example   left_end_read    1000   1500    .   +   .   BAC b101.2
ctgA   example   right_end_read   19500  20000   .   -   .   BAC b101.2
</pre></blockquote>

<p>

This is the contents of <a
href="data_files/volvox5.gff">volvox5.gff</a>.  Go ahead and add this
into the database now.  To visualize this you will:

<ol>
  <li>Create a custom aggregator that explains the relationship
      between the three feature types.
  <li>Create a new stanza that uses the custom aggregator.
</ol>


<p>

To define the custom aggregator, open volvox.conf and add the
following to the aggregators line:

<blockquote class="example"><pre>
aggregators = match
	      processed_transcript
	      coding
              BAC{left_end_read,right_end_read/BAC}
</pre></blockquote>

<p>

The thing named <b>BAC{left_end_read,right_end_read/BAC}</b> is the
custom aggregator definition.  Its format is
<i>aggregator_name{subtype1,subtype2,subtype3.../main_type}</i>.  Here
we're defining an aggregator of type "BAC" which has subparts of type
"left_end_read" and "right_end_read" (separated by commas) and
top-level type of "BAC" (separated from the subparts by a slash).
Although it's not necessary to use the same name for both the main
feature type and the aggregator, it's often convenient to do so.

<p>

Now add the appropriate stanza to the bottom of volvox.conf:

<blockquote class="example"><pre>
[Clones]
feature      = BAC
glyph        = segments
bgcolor      = yellow
strand_arrow = 1
description  = 1
key          = Fingerprinted BACs
</pre></blockquote>

<p>

With this new track turned on, look at ctgA:1..24200.  It will show
that GBrowse has correctly picked up and rendered the relationship
between the whole BAC and its two end reads (Figure 11).

<blockquote>
<img src="figures/custom_aggregators1.gif"><br>
<i>Figure 11: The glyph produced by a custom BAC aggregator</i>
</blockquote>

<p>

For your convenience, the configuration file with all the
modifications made up through this point of the tutorial can be found
in <a href="conf_files/volvox3.conf">volvox3.conf</a>.

<hr>

<h2><a name="graph">11. Showing Quantitative Data</a></h2>

<p>

GBrowse can plot quantitative data such as alignment scores,
confidence scores from gene prediction programs, and microarray
intensity data.  The data can be displayed either with glyphs that
change color to indicate score levels (see the
"heterogeneous_segments", "graded_segments" and "redgreen_box"
glyphs), or using a general-purpose XY-plot glyph.

<p>

Congratulations, Affymetrix has built a transcriptional profiling chip
for Volvox!  There's now a transcriptional profile for volvox, with an
intensity reading every 100 bp across all of ctgA.  The simulated data
for this is in the file <a
href="data_files/volvox6.gff">volvox6.gff</a>, an excerpt of which is
shown here:

<blockquote class="example"><pre>
ctgA   affy   tlevel   1     100   281   .   .   Affy Expt1
ctgA   affy   tlevel   101   200   183   .   .   Affy Expt1
ctgA   affy   tlevel   201   300   213   .   .   Affy Expt1
ctgA   affy   tlevel   301   400   191   .   .   Affy Expt1
ctgA   affy   tlevel   401   500   288   .   .	 Affy Expt1
...
</pre></blockquote>

<p>

The file contains 500 features, each of which is exactly 100 bp long.
The features are of type "tlevel" ("transcriptional level") and of
source "affy."  Each one has a score (column 6) between 0 and 1000,
where higher scores means more transcriptional activity.  This is the
first time we've used the score column.

<p>

All of the 500 features share the same group (column 9) of "Affy
Expt1."  They are grouped in this way because the entire set of 500
features represents a single transcriptional profiling experiment.  If
we had multiple experiments to show, they would be named Expt1, Expt2
and so on.

<p>

We would like to generate a line graph that shows the transcriptional
profile level across the current region.  To do this, we will first
create an aggregator that will bring all the individual tlevel
features together into a single feature named "tprofile."  This is
done as described in the previous section.  Modify the configuration
file's aggregators option to read as follows:

<blockquote class="example"><pre>
aggregators = match
	      BAC{left_end_read,right_end_read/BAC}
	      processed_transcript
	      coding
	      tprofile{tlevel}
</pre></blockquote>

The last line is declaring an aggregated feature named "tprofile"
whose parts consist of individual "tlevel" features.  This is similar
to the <a href="#custom_aggregators">BAC aggregator</a>, except that
in this case there is no top-level feature that goes from end to end,
so we just leave out the <i>/main_type</i> part of the aggregator
definition.

<p>

We now need to use this aggregated feature in a track stanza.  Create
the following section:

<blockquote class="example"><pre>
[TransChip]
feature        = tprofile
glyph          = xyplot
graph_type     = line
height         = 50
min_score      = 0
max_score      = 1000
scale          = right
key            = Transcriptional Profile
</pre></blockquote>

<p>

The options shown here create a track named TransChip to display the
tprofile feature with the xyplot glyph.  The "graph_type", "height",
"scale", "min_score", and "max_score" options all configure various
aspects of the xyplot glyph's appearance.

<blockquote class="example"><i>You can read all about xyplot's options using
<b>perldoc Bio::Graphics::Glyph::xyplot</b></i></blockquote>

<p>

When you reload the page and turn on the Transcriptional Profile
track, you should see something like that shown in Figure 12.

<blockquote>
<img src="figures/graph1.gif"><br>
<i>Figure 12: A transcriptional profile rendered with the xyplot glyph</i>
</blockquote>

<p>

<blockquote class="example"> <i>Using the info that perldoc provides, play around with
the xyplot options a bit.  For example, see what happens when you
change graph_type to "boxes."</i> </blockquote>

<hr>

<h2><a name="dna">12. Displaying DNA and 3-frame translations</a></h2>

<p>

GBrowse can take advantage of DNA sequence data in several ways:

<ol>
  <li>It can display a GC content graph of the reference sequence at
      low magnifications and the DNA sequence itself at higher
      magnifications.
  <li>It can display three and six-frame translations of the reference
      sequence DNA.
  <li>It can display the protein translation of coding regions.
  <li>It can display aligned nucleotide sequences, creating a poor
      man's multiple alignment.
</ol>

<p>

So we've been working with feature coordinates, but no actual DNA
sequence has been loaded into the volvox database.  We will again
rebuild the database, this time loading in a simulated DNA file in
fasta format.  Download the file <a
href="data_files/volvox.fa">volvox.fa</a>, and copy it into the volvox
database directory.  At this point in the tutorial, when you do a
directory listing of the volvox database directory (with "ls" on
unix systems, or "dir/w" on Windows systems) it should look like
this:

<blockquote class="example"><pre>
% <b>ls /tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases/volvox/</b>
volvox.fa    volvox2.gff  volvox4.gff  volvox6.gff
volvox1.gff  volvox3.gff  volvox5.gff
</pre></blockquote>

<p>

After copying the .fa file into the volvox database directory, you
will need to change the configuration file very slightly to tell
GBrowse to look for and load the FASTA file.  At the top of the config
file, change the <tt>db_args</tt> section to look like this:

<p>

<pre>
db_args       = -adaptor memory
		-dir     '/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases/volvox'
</pre>

<p>

Previously the <tt>-gff</tt> argument told GBrowse to load all GFF
files in the directory.  The new <tt>-dir</tt> argument says to load
both GFF files and FASTA sequence files.  (There's also a
<tt>-fasta</tt> argument to load sequence without features, but this
is not much use with GBrowse.)

<p>

This is all you need to do to load the DNA.  To see that the DNA is
indeed being loaded, add two new stanzas to the volvox.conf
configuration file:

<blockquote class="example"><pre>
[DNA]
glyph          = dna
global feature = 1
height         = 40
do_gc          = 1
fgcolor        = red
axis_color     = blue
strand         = both
key            = DNA/GC Content

[Translation]
glyph          = translation
global feature = 1
height         = 40
fgcolor        = purple
start_codons   = 0
stop_codons    = 1
translation    = 6frame
key            = 6-frame translation
</pre></blockquote>

<p>

The "DNA" track uses a specialized glyph called "dna".  At low
magnifications (zoomed way out), this glyph draws a GC content plot.
At high magnifications (zoomed way in), this glyph draws the dna.  Of
the various options given in the example stanza, the most important
one is "global feature", which is set to a true value (1).  This tells
GBrowse that the stanza doesn't correspond to a specific feature type,
but should be displayed globally.  Other options control whether to
draw one or both strands, whether to draw the GC content histogram,
and what colors to use.

<p>

Similarly, the "Translation" track uses a glyph called "translation",
which draws three or six-frame conceptual translations.  At low
magnifications (zoomed way out), this glyph draws little symbols
indicating where start and stop codons are.  At high magnifications,
the actual amino acid sequence comes into view. Again, the most
important option is "global feature", which is set to a true value to
tell GBrowse that the track isn't attached to a particular feature
type, but is to be generated automatically.  Other options control the
height of the glyph, whether to draw start and/or stop codon symbols,
and whether to generate a 3frame or 6frame translation.

<p>

Figures 13a and 13b show the browser at low and high magnification, with
both tracks activated.  Notice that the coding track ("cds" glyph)
notices that the DNA is available and generates the transcripts'
protein translations automatically!

<blockquote>
<b>(13A)</b><br>
<img src="figures/dna1.gif">
<br><br><b>(13B)</b><br>
<img src="figures/dna2.gif"><br>
<i>Figure 13: Viewing DNA/GC content and 6-frame translation. (a) low
magnification; (b) high magnification</i>
</blockquote>

<p>

<blockquote><i>
If you happen to do a listing of the volvox database directory after
adding the DNA file, you might notice that a new file named
"directory.index" has appeared.  This index directory is created
automatically by GBrowse in order to speed up access to the .fa file
and to reduce memory requirements.  If the database directory is
<b>not</b> writable by all users, GBrowse will not be able to create
this directory, and the display will be somewhat slower whenever a
DNA track is turned on.
</i></blockquote>

<hr>

<h2><a name="multiple_alignments">13. Displaying ESTs and Other
Alignments</a></h2>

<p>

This section will lead you through creating a plausible EST track, and
show you how grouping of 5' and 3' EST reads works.

<p>

We'll start with a simple data set containing information on three
pairs of EST reads.  You'll find this data set in <a
href="data_files/volvox7.gff">volvox7.gff</a>.  Here is the first pair
described in the data file:

<blockquote class="example"><pre>
ctgA   est   match 1050   3202   .   +   .   EST agt830.5
ctgA   est   HSP   1050   1500   .   +   .   EST agt830.5
ctgA   est   HSP   3000   3202   .   +   .   EST agt830.5
ctgA   est   match 5410   7503   .   -   .   EST agt830.3
ctgA   est   HSP   5410   5500   .   -   .   EST agt830.3
ctgA   est   HSP   7000   7503   .   -   .   EST agt830.3
...
</pre></blockquote>

<p>

What's going on here is the same as the alignments shown in <a
href="data_files/volvox3.gff">volvox3.gff</a>.  There are two EST
reads named agt830.5 (the 5' read) and agt830.3 (the 3' read).  Each
of them matches the ctgA genome in two discontinuous regions because,
presumably, they cross a splice site.  As in the earlier example, we
represent each EST as a single "match" feature that spans the entire
region, plus a series of "HSP" features that correspond to the aligned
regions.  The last column is used to group the match and HSP features
together using the class and name of the feature, where the class is
arbitrarily chosen to be "EST."

<p>

There are two other things to notice.  One is that the source field
(column 2) is "est".  All previous examples used "example" here.  This
is because we need to distinguish this set of alignments from the
generic alignments in <a
href="data_files/volvox3.gff">volvox3.gff</a>.  The second item of
interest is that the strand field (column 7) is + for the 5' EST and -
for the 3' EST, indicating that the 3' EST aligned to the reverse
complement of ctgA.

<p>

Add this file to the volvox database directory, and add the following
to the configuration file:

<blockquote class="example"><pre>
[EST]
feature      = match:est
glyph        = segments
height       = 6
bgcolor      = orange
key          = ESTs
</pre></blockquote>

<p>

This will give a display similar to that shown in Figure 14.

<blockquote>
<img src="figures/multiple_alignments1.gif"><br>
<i>Figure 14:  A simple representation of EST matches.</i>
</blockquote>

<p>

Notice that the feature option reads "match:est" rather than simply
"match."  This is to distinguish the EST matches from the example
matches that we loaded previously.  When needed, you can use the
source field (column 2) to distinguish different features of the same
type, using the format "type:source".  You can use this equally well
with ordinary types (e.g. "my_feature:example") or with aggregated
types ("processed_transcript:genscan").

<p>

This display is OK, but it could be better.  One problem is that the
relationship between the 5' and 3' EST read pairs is not shown.  We'd
like to place the two members of the pair together on the same line,
and connect them with a dotted line to show that they are the two ends
of the same cDNA clone.  Recall that we did something similar to this
with the custom <a name="custom_aggregators">BAC aggregator</a>.
Unfortunately, there's a problem with the ESTs because we are already
using the "match" aggregator to perform one level of grouping, and the
GFF load format only allows one level of grouping at a time (this is
changing in a proposed new version of the format).

<p>

For the time being, we can work around this problem using a "hack."
Change the [EST] track configuration to look like this:

<blockquote class="example"><pre>
[EST]
feature       = match:est
glyph         = segments
bgcolor       = orange
group_pattern = /\.[53]$/
key           = ESTs
</pre></blockquote>

<p>

The new group_pattern option tells GBrowse to use a Perl regular
expression pattern matching operation to find and group related EST
matches based on their names.  It helps to understand how Perl regular
expressions work, but basically the pattern match breaks down this
way:

<blockquote class="example"><pre>
  /            begin the pattern match
  \.           match a dot
  [53]         match either the numbers 5 or 3
  $            match the end of the string
  /            end the pattern match
</pre></blockquote>

<p>

What this is saying is to look for pairs of EST names that are similar
except for the terminal .5 or .3, and pair them.  When we reload the
page, we get Figure 15.

<blockquote>
<img src="figures/multiple_alignments2.gif"><br>
<i>Figure 15: The group_pattern option allows EST pairs to be grouped</i>
</blockquote>

<p>

Here are regular expressions that will work for other common EST
pairing schemes:

<table border="1">
  <tr>
    <th>5' EST</th>
    <th>3' EST</th>
    <th>group_pattern</th>
  </tr>
  <tr>
    <td>agt123f</td>
    <td>agt123r</td>
    <td>/[fr]$/</td>
  </tr>
  <tr>
    <td>agt123p</td>
    <td>agt123q</td>
    <td>/[pq]$/</td>
  </tr>
  <tr>
    <td>f.agt123</td>
    <td>r.agt123</td>
    <td>/^[fr]\./</td>
  </tr>
  <tr>
    <td>5.agt123</td>
    <td>3.agt123</td>
    <td>/^[53]\./</td>
  </tr>
  <tr>
    <td>agt123.for</td>
    <td>agt123.rev</td>
    <td>/\.(for|rev)$/</td>
  </tr>
</table>

<p>

Another nice enhancement would be to give the 5' and 3' ESTs different
colors so as to distinguish one from another.  This can be
accomplished using a Perl <i>callback</i>.  Open up volvox.conf once
more, and find the bgcolor option in the [EST] track.  Replace it with
this (you may want to cut and paste from here in order to avoid
introducing any typos):

<blockquote class="example"><pre>
bgcolor      = sub {
		my $feature = shift;
		my $name    = $feature->display_name;
		if ($name =~ /\.5$/) {
		   return 'red';
		} else {
		   return 'orange';
		}
	}
</pre></blockquote>

<p>

You'll need to know the basics of the Perl programming language in
order to do this type of thing yourself.  Suffice to say that instead
of hard-coding the color "orange" into the bgcolor option, we are
asking GBrowse to run a Perl subroutine each time it needs to render
an EST.  The subroutine is passed the feature that is about to be
drawn.  It asks the feature for its human-readable name (display_name)
and assigns that name to a variable named $name.  It then performs a
pattern match on the name to see if it ends in a "5".  If the name
matches, the subroutine returns the color "red" to GBrowse.  Otherwise
it returns the color "orange."

<p>

The effect is shown in Figure 16.

<blockquote>
<img src="figures/multiple_alignments3.gif"><br>
<i>Figure 16: Using a callback to distinguish 5' and 3' ESTs</i>
</blockquote>

<hr>

<h2><a name="adding_dna_to_alignments">14. Adding DNA to Alignments</a></h2>

<p>

The last thing we'll do with the EST data set is to add DNA to the
ESTs so that at high magnification GBrowse will show the multiple
alignment.  This information is also used by the "dump alignments"
plugin to generate a text-based multiple alignment.

<blockquote class="example"> <i> NOTE: Currently only nucleotide to nucleotide
alignments can be displayed at the level of individual nucleotides
(e.g. BLASTN, BLAT, Exonerate).  Protein to nucleotide alignments,
such as those produced by Genewise or BLASTX, are not supported at the
residue level</i>
</blockquote>

<p>

To make this work, we need to add two additional pieces of information
to the EST alignment data:

<ol>
  <li>The DNA sequences of the volvox ESTs.
  <li>The alignment positions in EST coordinates.
</ol>

In case the need for item (2) isn't immediately clear, consider this
blow-up of an alignment:

<blockquote class="example"><pre>
ctgA      1050 gattgccattgaccttggccattggccaagctgaa 1086
               |||||||||| ||||||| ||||||||||||||||
agt830.5     1 gattgccattcaccttgggcattggccaagctgaa 135
</pre></blockquote>

<p>

What we currently have in the GFF file are the <b>source</b> genomic
positions of the alignments (in ctgA-relative coordinates).  We need
to add the <b>target</b> positions in agt830.5-relative coordinates in
order for GBrowse to fetch and display the appropriate segments of the
EST DNA.

<p>

The fasta file <a href="data_files/ests.fa">ests.fa</a> provides the DNA
sequences for the six EST reads.  The GFF load file
<a href="data_files/volvox8.gff">volvox8.gff</a> contains the revised
coordinates.  If you look at this file you'll see that it is
dissimilar to previous load files:

<blockquote class="example"><pre>
ctgA	est	match	1050	3202	.	+	.	Target EST:agt830.5 1 554
ctgA	est	HSP	1050	1500	.	+	.	Target EST:agt830.5 1 451
ctgA	est	HSP	3000	3202	.	+	.	Target EST:agt830.5 452 654
ctgA	est	match	5410	7503	.	-	.	Target EST:agt830.3 1 595
ctgA	est	HSP	5410	5500	.	-	.	Target EST:agt830.3 505 596
ctgA	est	HSP	7000	7503	.	-	.	Target EST:agt830.3 1 504
</pre></blockquote>

<p>

The first eight columns are identical to what we've been using before,
but the ninth column follows a new convention used for nucleotide to
nucleotide and protein to nucleotide alignments.  There is now a
special class name, "Target", that tells GBrowse that the group field
represents the combination of a target sequence and its coordinates.
Following Target is the name EST:agt830.5, which is a composite of the
"real" class name ("EST") and the name of the EST.  The two are
separated by a colon in the format "class:name".  Following this are
two numbers indicating the start and end of the alignment in EST
coordinates.

<p>

There are a couple of subtleties to notice here.  First of all, notice
that the "match" features extend all the way across the matched area
of the genome (1050 to 3202 in the case of agt830.5), and all the way
across the matched area of the target (1 to 554). Because one or both
of the matched regions may contain gaps, the source and target regions
do not have to be the same length.  On the other hand, each HSP covers
an ungapped contiguous region: the first HSP covers 1050..1500 in
genome coordinates and 1..451 in target (EST) coordinates; the second
covers 3000..3202 in genome coordinates and 452..654 in target
coordinates.  The HSPs should have the same alignment length in both
genome and target coordinates, or at least very close lengths.  If the
lengths are close but not identical, GBrowse will realign the
segments, introducing small gaps where necessary.

<blockquote class="example"> <i>NOTE: GBrowse uses a simple but slow segment
realigner.  If there are numerous gaps, it is better to break them
into a set of smaller colinear HSPs than to rely on the realigner to
do it for you.</i> </blockquote>

<p>

The second subtlety to notice is that for the minus strand ESTs, the
target coordinates are not reversed, that is, the start position is
less than the end position.  For example, for the first agt830.3
HSP, we are told that genomic region 5410..5500 aligns to EST region
505..596.  The strand field is used to determine the direction of the
alignment.

<blockquote class="example"><i>Note that this contradicts the historical
implementation of GFF but the current use is more internally consistent and is 
the method for target annoation in the proposed revision of GFF.
</i></blockquote>

<p>

Since this data file contains a revised version of volvox7.gff,
<b>remove volvox7.gff from the database directory and replace it with
<a href="data_files/volvox8.gff">volvox8.gff</a> </b>.  Also copy <a
href="data_files/ests.fa">ests.fa</a> into the database directory.  If
you perform a directory listing, it should look like this:

<blockquote><pre>
directory.index  volvox.fa    volvox2.gff  volvox4.gff	volvox6.gff
ests.fa		 volvox1.gff  volvox3.gff  volvox5.gff	volvox8.gff
</pre></blockquote>

<p>

<blockquote> <i>NOTE: If you see doubled EST features after this
point, make sure that you have removed volvox7.gff.  Another thing to
watch out for is that some sort of bug in the BioPerl layer (up
through at least version 1.4) causes the EST DNA display to get messed
up at this point on Windows systems.  To fix the latter problem, go to
the volvox database directory and remove the files directory.dir and
directory.pag.  These are automatically-generated DNA file indexes
that GBrowse develops, and will be regenerated for you the next time
you access a page.</i> </blockquote>


<p>

We're not done with making configuration file changes, but <a
href="conf_files/volvox4.conf">volvox4.conf</a> contains all
configuration file enhancements up to this point.  If you like, you
can copy it over the live volvox.conf.  It contains the following
version of the [EST] track:

<blockquote class="example"><pre>
[EST]
feature      = match:est
glyph        = segments
height       = 6
draw_target   = 1
show_mismatch = 1
canonical_strand = 1
bgcolor      = sub {
		my $feature = shift;
		my $name    = $feature->display_name;
		if ($name =~ /\.5$/) {
		   return 'red';
		} else {
		   return 'orange';
		}
	}
group_pattern = /\.[53]$/
key          = ESTs
</pre></blockquote>

<p>

The key addition to this track configuration is the "draw_target",
"show_mismatch" and "canonical_strand" options.  All options are
true/false flags, where 0 means false and 1 means true.  draw_target
tells the segments glyph to draw the DNA sequence of the target ESTs
when the magnification allows.  show_mismatch instructs the glyph to
highlight mismatches between the genome and the EST in pink.
canonical_strand instructs the glyph to display the plus strand
sequence even when the EST matches the minus strand.

<p>

To see this work, reload the page, turn on the EST track and search
for region "ctgA:1065..1165".  This will show the aligned 5' ends of
agt221.5, agt830.5 and agt767.5 (Figure 17).  Notice that one of the
T's towards the beginning of agt830.5 is highlighted in red, to show
that it doesn't match the corresponding genomic base.

<blockquote>
<img src="figures/adding_dna_to_alignments1.gif"><br>
<i>Figure 17: Multiple alignments at the DNA level</i>
</blockquote>

<p>

If you don't see the EST sequence appearing, make sure that ests.fa is
in the volvox database directory and is world readable.  If it still
isn't working, you may need to "touch" the file in order to update its
modification date.  This tells GBrowse that it is new and needs to be
reindexed.  In Unix:

<blockquote class="example"><pre>
% <b>touch /tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases/volvox/ests.fa</b>
</pre></blockquote>

<hr>

<h2><a name="trace">15. Displaying Trace Data</a></h2>

<p>

If you have sequence trace information (in SCF format) associated with
the reference sequence, this can be displayed in gbrowse using the
trace glyph. To use this glyph, you must have installed:

<dl>
  <dt>The Staden io-lib package
  <dd><a href="http://staden.sourceforge.net">staden.sourceforge.net</a>
  <dt>zlib
  <dd><a href="http://www.zlib.net">www.zlib.net</a>
  <dt>The Bio::SCF perl module
  <dd>Available from CPAN
</dl>

Note that at this time, it is not possible to use the trace glyph
with Windows servers, since we do not know of a version of the Staden
io-lib package that has been compiled for Windows.
<p>

The data file <a href="data_files/volvox9.gff">volvox9.gff</a> contains an example trace entry.

<p>

<blockquote class="example"><pre>
ctgA    example trace   44401   45925   .       +       .       name trace; trace volvox_trace.scf
</pre></blockquote>

<p>

This aligns the full trace sequence to the reference sequence.  The trace file in this case is named "volvox_trace.scf". 

Due to sequence quality, the first few bases of a trace file usually don't align.  Even so, these need bases need to be included in the gff file.  For instance, if the bases 10-700 of the trace file aligns to the bases 100-800 of the reference sequence, the feature would be 90-800 to account for the first 10 bases (starting at base 0).

<p>

<blockquote class="example"> <i> NOTE: The trace glyph currently doesn't deal with insertions or deletions.  If an indel occurs, the alignment after the indel will be off.</i>
</blockquote>

<p>

To display this first copy the following into the volvox.conf (or copy <a href="conf_files/volvox5.conf">volvox5.conf</a> over the volvox.conf file).

<p>

<blockquote class="example"><pre>
[Traces]
feature      = trace
glyph        = trace
fgcolor      = black
bgcolor      = orange
strand_arrow = 1
height       = 6
description  = 1
a_color      = green
c_color      = blue
g_color      = black
t_color      = red
trace_height = 80
trace_prefix = http://localhost/gbrowse/tutorial/data_files/
show_border  = 1
key          = Traces
</pre></blockquote>

<p>

The fgcolor, bgcolor, strand_arrow and height control the bar that shows the location and directionality of the trace.

<p>

<p>

The trace_prefix option is important because it gives the path to the trace files.  This is prepended to the trace file name defined in the gff file.  It can be a direct path to the directory (eg "/usr/local/trace_files/") or a web address (as above). 

<p>

The a/c/g/t_color options allow configuration of the base colors.  The trace_height refers to the height of the trace itself.  Play around with it to find a height that you like.

<p>

If show_border is set to 1, a black box will be drawn around the trace.

<p>

After configuring the trace glyph, reload the browser page and enable traces.  Zoomed out you will see:

<p>

<blockquote>
<img src="figures/trace1.png"><br>
<i>Figure 18: The trace glyph zoomed out.</i>
</blockquote>

<p>

Zooming in will show you the trace diagram:

<p>

<blockquote>
<img src="figures/trace2.png"><br>
<i>Figure 19: The trace glyph zoomed in.</i>
</blockquote>

<p>

<hr>

<h2><a name="plugins">16. Plugins</a></h2>

<p>

Is all the effort to load the genomic and EST DNAs worth it?  Yes, if
you want to take advantage of two popular plugins,
<i>RestrictionAnnotator</i> and <i>Aligner</i>.  The first generates a
track of restriction sites.  The second dumps a text-based multiple
alignment of the current region on view.

<p>

To see these plugins at work, first make sure that the database files
are up to date with this position in the tutorial.  If you are in any
doubt, remove the current contents of the volvox database directory
and replace them with the files <a
href="data_files/volvox_all.gff">volvox_all.gff</a> and <a
href="data_files/volvox_all.fa">volvox_all.fa</a>.  <p>

Now find the option "plugins=" at the top of volvox.conf, and modify
it to activate the Aligner and RestrictionAnnotator plugins:

<blockquote class="example"><pre>
plugins = Aligner RestrictionAnnotator
</pre></blockquote>

<p>

When you reload the page, you will see a new popup menu appear under
the image labeled "Dumps, searches and other operations."  You will
also see an automatic track labeled "plugin:Restriction Sites" appear
in the track list.  When you turn on this track, you will be presented
with a restriction map (Figure 18).  You can then adjust which
restriction sites are shown by selecting "Annotate Restriction Sites"
from the popup menu and pressing the "Configure" button.

<blockquote>
<img src="figures/plugins1.gif"><br>
<i>Figure 20: The RestrictionAnnotator Plugin</i>
</blockquote>

<p>

To see the Aligner at work, center your view on a region that contains
the EST alignments (for example, ctgA:1000..5000), select "Dump
Alignments" from the plugin popup menu, and press "Go".  This will
return a text-based multiple alignment of the genome and the EST
tracks.

<p>

The Aligner plugin has some additional configuration that you can
perform.  We'll look at this now as an example of how to configure
plugins. Open up volvox.conf and add the following configuration
section:

<p>

<blockquote class="example"><pre>
########################
# Plugin configuration
########################

[Aligner:plugin]
alignable_tracks   = EST
upcase_tracks      = CDS Motifs
upcase_default     = CDS
</pre></blockquote>

<p>

It doesn't matter where the section goes, but it is probably a good
idea to place this towards the middle of the file after the [GENERAL]
section (at the top) and before the [TRACK DEFAULTS] section.
Otherwise it is easy for you or someone else maintaining the
configuration file to mistake this for some sort of track
configuration.

<p>

Plugin configuration sections are distinguished from track
configuration by having names of the format
<b><i>PluginName</i>:plugin</b>.  In this case, the three
configuration options are applied to the Aligner plugin.  For the
Aligner plugin, the configuration options are:

<table border="1">
  <tr>
    <th>Option</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>alignable_tracks</td>
    <td>Space-delimited list of tracks to include in the multiple
	alignment.  The genome is always included.  If this option is
	not present, then GBrowse will automatically include any track
	that has the "draw_target" option set.
    </td>
  </tr>
  <tr>
    <td>upcase_tracks</td>
    <td>Space-delimited list of tracks that will be used to UPCASE the
	genomic DNA.  This is very useful if you want to embed the
	positions of coding regions or other features inside the
	multiple alignment.  Uppercasing will not be turned on by
	default.  The user must press the "Configure" button, and
	select which of the uppercase tracks are to be activated  from
	a list of checkboxes.
    </td>
  </tr>
  <tr>
    <td>upcase_default</td>
    <td>A space-delimited list of tracks that will be uppercased by
	default unless the user turns them off during configuration.
    </td>
  </tr>
  <tr>
    <td>ragged_default</td>
    <td>A small integer indicating that the aligner should include
	some unaligned bases from the end of each sequence.  This is
	useful for seeing the sequencing primer or cloning site in ESTs.
    </td>
  </tr>
</table>

<p>

With the changes in place, select the aligner from the popup menu and
press Configure.  Turn on uppercasing of the coding region track and
see how it affects the display (Figure 19).

<blockquote>
<img src="figures/plugins2.gif"><br>
<i>Figure 21: The Aligner plugin produces multiple alignments.</i>
</blockquote>

<p>

Only a few of the plugins are currently well-documented, but this
situation is being rectified.  To view their documentation, if any,
find the plugin file, which usually lives under gbrowse.conf/plugins,
and run the perldoc command with the -F ("file") option:

<blockquote class="example"><pre>
% <b>perldoc -F Aligner.pm</b>
</pre></blockquote>

<p>

Here's the list of plugins that come with the standard distribution:

<table border="1">
  <tr>
    <th>Plugin</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>Aligner</td>
    <td>Dump multiple alignments</td>
  </tr>
  <tr>
    <td>BatchDumper</td>
    <td>Allows the user to cut and paste a series of landmarks on the
	genome and dumps out all overlapping features using a variety
	of formats (e.g. GenBank format)
    </td>
  </tr>
  <tr>
    <td>FastaDumper</td>
    <td>Produce pretty-printed FASTA dumps of the current region, with
	selected features highlighted with colors or font styles.
    </td>
  </tr>
  <tr>
    <td>GFFDumper</td>
    <td>Dump out the current region in GFF format (redundant with BatchDumper).</td>
  </tr>
  <tr>
    <td>OligoFinder</td>
    <td>Lets the user search for landmarks on the basis of unique
	11-mers or greater.</td>
  </tr>
  <tr>
    <td>RestrictionAnnotator</td>
    <td>Creates restriction maps.</td>
  </tr>
  <tr>
    <td>SequenceDumper</td>
    <td>The same functionality as BatchDumper, but just shows features
	that overlap the current region on view.</td>
  </tr>
</table>

<hr>

<h2><a name="reference_sequence">17. More about the Reference Sequence</a></h2>

<p>

Although the example that we've been working with only has a single
reference sequence (the infamous "ctgA"), many projects will have
multiple references.  Reference sequences can be anything that acts as
a convenient landmark: sequenced clones, contigs, scaffolds, golden
path segments, or whole chromosomes.

<p>

There are just a few rules to be aware of when setting up the load
files:

<ol>
  <li>All reference sequences must share the same <b>class</b> name.  In our
      example, the class for the reference sequence was <i>Contig</i>.
  <li>Each reference sequence must have its own feature entry in the GFF load
      file.  This feature entry must use itself as the reference
      sequence, start at position 1, and extend the length of the
      sequence. The source and type for the feature are arbitrary, and
      do not have to be the same across all reference sequences.
  <li>The class chosen for reference sequences must be noted in the
      configuration file under the general option "reference class."
</ol>

<p>

Let us review these three criteria for the volvox example.  If you
look at the top of the initial load file, <a
href="data_files/volvox1.gff">volvox1.gff</a>, you'll see the very
first line is:

<blockquote class="example">
<pre>
ctgA   example   contig       1      50000   .   .   .   Contig ctgA
</pre>
</blockquote>

<p>

1. The class for the name "ctgA" is <i>Contig</i>.  If we were using
other reference sequences as landmarks, then they too would have to be
identified as Contigs.

<p>

2. This line describes "ctgA" as a feature relative to itself.  The
feature starts at position 1 (it has to!) and ends at position 50,000.
The source is "example" and the type is "contig."  If there are other
reference sequences in the database, they do <b>not</b> have to share
the "contig" type.  This allows you refer to other types of landmarks,
such as clones.

<p>

3. If you examine the top few lines of the volvox.conf configuration
file, you'll see this line:

<blockquote class="example"><pre>
reference class  = Contig
</pre></blockquote>

<p>

This line is required for GBrowse to effectively find and render with
features located on the reference sequence.

<p>

If you find this confusing, it might help to choose "Reference" as the
class.  Then you can write GFF load files like this:

<blockquote class="example">
<pre>
ctgA   example   contig       1      50000   .   .   .   Reference ctgA
chr22  example   chromosome   1    1150000   .   .   .   Reference chr22
5p3.2  example   band         1     830000   .   .   .   Reference 5p3.2
</pre>
</blockquote>

<blockquote class="example"><i>Don't forget to update "reference class" in the config
file!</i>
</blockquote>

<p>

Alternatively, you can use the word "Sequence" as the class name for
reference sequences.  For historical reasons, "Sequence" is recognized
as the default classname.  This means you don't have to have a
"reference class" option in the config file at all.

<hr>

<h2><a name="semantic_zooming">18. Semantic Zooming</a></h2>

<p>

One of the cooler features of GBrowse is its ability to support
semantic zooming.  Semantic zooming is a feature in which objects show
different levels of detail depending on the level of magnification.
We've already seen this behavior in the "dna" and "segments" glyphs,
which show the DNA sequence only when there's sufficient room to
display it.

<p>

GBrowse has several types of semantic zooming:

<dl>
  <dt>glyph-based, automatic
  <dd>The dna and segments glyphs, and others that support semantic
      zooming out of the box.  This happens automatically and can't
      be modified.
  <dt>semantic labeling
  <dd>When there's sufficient room, GBrowse will print the label and
      descriptions next to the glyphs.  The threshold at which this
      happens is under your control.
  <dt>semantic bumping
  <dd>When there's sufficient room, GBrowse will "bump" features to
      prevent them from colliding on the screen.  When this would
      cause the display to become to high, bumping is suppressed.
      This threshold is also under your control.
  <dt>semantic options
  <dd>You can set track configuration sections up so that when a
      preset size threshold is exceeded, one configuration replaces
      another.
</dl>

<p>

The thresholds for labeling and bumping are set by configuration
options named "label density" and "bump density" respectively.  The
standard values can be found in the defaults track named [TRACK
DEFAULTS].  They are originally set so that labels are suppressed when
there are more than 25 features per track, and bumping is suppressed
when there are more than 100 features per track.  You can these values
globally by editing their values in [TRACK DEFAULTS], or you can add
"label density" and/or "bump density" options to individual track
configuration sections in order to override the settings for specific
tracks.

<p>

The process of setting up semantic options is a bit more interesting.
To illustrate, we will create semantic zooming for the [Alignments]
track ("Example Alignments").  We would like the track to shift from
showing the individual segments to showing solid rectangles when the
user is zoomed out to 30K and beyond, and turn bumping off when the
user is zoomed out to 45K and beyond.  The process is simple.  Beneath
the [Alignments] stanza, we add a stanza qualified for zoomlevels of
>= 30,000 and another stanza qualified for zoomlevels of >= 45,000:

<blockquote class="example"><pre>
[Alignments]
feature      = match
glyph        = segments
key          = Example alignments

[Alignments:30000]
glyph        = box
label        = 0

[Alignments:45000]
glyph        = box
bump         = 0
label        = 0
</pre></blockquote>

<p>

The format for semantic options is [<i>Trackname:distance</i>], where
<i>Trackname</i> must be the same as the non-qualified track, and
<i>distance</i> is the length of the region at which the semantic
options will kick in.  Only options that are different from the
non-qualified track need to be listed.  According to the configuration
given above, when the user is looking at a region 30,000 bp or longer,
the glyph option will change to "box," which is a solid rectangle that
doesn't show any internal details.  All other options, such as feature
and key, will be inherited from the [Alignments] track.

<p>

At 45,000 bp, the glyph is again set to box, and in addition the
"bump" option is set to zero, turning off collision control.  Notice
that options are inherited from the unqualified track stanza, and not
from the previous semantic zoom level.  If we had neglected to specify
the glyph option in [Alignments:45000], the glyph would have reverted
to "segments."

<p>

Make these changes to volvox.conf, turn on the "Example Alignments"
track, and view the contig at 20K, 40K and 50K.  At 40K, you'll see
the alignments lose their internal structure and be replaced by solid
boxes (Figure 20).  At 50K they'll begin to overlap and the feature
labels will be suppressed.

<blockquote>
<img src="figures/semantic_zooming1.gif"><br>
<i>Figure 22: Semantically zoomed alignments at 40K</i>
</blockquote>

<hr>

<h2><a name="overview">19. Putting Features into the Overview</a></h2>

<p>

The overview is the scale that appears at the top of the detailed
image.  In addition to acting as a reference point and navigation
tool, you can place tracks in it.  These tracks will always be
displayed, and can serve as reference points for well-known genes,
cytogenetic bands, or genetic markers.

<p>

We will illustrate how to do this by placing a copy of the Motifs
track into the overview.  Add the following to the bottom of the
volvox.conf configuration file:

<blockquote class="example"><pre>
[Motifs:overview]
feature      = motif
glyph        = span
height       = 5
description  = 0
label        = 1
key          = Motifs
</pre></blockquote>

<p>

This stanza is identical to the [Motifs] track that we created
earlier, except that its name is qualified with ":overview".  This
tells GBrowse that this is not an ordinary track to be placed in the
detail image, but one that should be placed in the overview.

<p>

We also want the overview motifs track to be displayed by default, so
go to the top of the configuration file, and modify the "default
features" option to look like this:

<blockquote class="example"><pre>
# list of tracks to turn on by default
default features = ExampleFeatures  Motifs:overview
</pre></blockquote>

<p>

Reload the page.  Viol&aacute;!  See Figure 21.

<blockquote>
<img src="figures/overview1.gif"><br>
<i>Figure 23:</i> Any number of tracks can be placed in the overview
</blockquote>

<p>

You can add as many tracks to the overview as you like.  The main
warning is that if you add lots of features to the overview it can get
pretty crowded in there.  Performance can also suffer, since each
feature must be fetched and rendered each time the overview is
displayed.

<hr>

<h2><a name="linking">20. Linking</a></h2>

<p>

The next topic we'll cover in this tutorial is configuring GBrowse's
outgoing links.  When the user clicks on a glyph in the details image,
he will be taken to another page by following a URL.  The URL to
follow is generated from the <code>link</code> option.  The default
link option is located in the [TRACK DEFAULTS] section of the config
file; you can specify track-specific links by placing a
<code>link</code> option in one or more of the individual track
stanzas.

<p>

The volvox.conf track defaults looks like this:

<blockquote class="example"><pre>
[TRACK DEFAULTS]
glyph         = generic
height        = 10
bgcolor       = lightgrey
fgcolor       = black
font2color    = blue
label density = 25
bump density  = 100
# where to link to when user clicks in detailed view
link          = AUTO
</pre></blockquote>

<p>

In this case, we've been using a special link URL of "AUTO."  This
generates an automatic link to a helper script named
"gbrowse_details."  If you click on some of the features in the
current volvox page you'll get an idea of what this script displays.
Try clicking on a motif, a spliced transcript, the EDEN gene, and an
EST.  When you click on the spliced transcript, notice that the
content of the "Gene" attribute is displayed.  By adding attributes
like this one, you can build up a very modest web-browsable database
of facts about your features.

<p>

We're going to override the default link rule for the motif track.
There's nothing sensible to link to, so we'll link to Google using
first the motif's name, and then the motif's description.

<p>

Go to the [Motifs] stanza in the volvox.conf config file and modify it
so that it looks like this:

<blockquote class="example"><pre>
[Motifs]
feature      = motif
glyph        = span
height       = 5
description  = 1
link         = http://www.google.com/search?q=$name
key          = Example motifs
</pre></blockquote>

<p>

The only change we've made is to add a "link" option to the stanza,
where the value is a Google search URL.  "$name" is a Perl variable.
GBrowse will fill in this variable with the name of the motif.  Reload
the page and click on a motif to see that this works as advertised
("m01," "m02" and the other example motifs are similar to the names
for galactic clusters, so be prepared for some astronomy hits).

<p>

It would be more sensible to link to the description of the motif, for
example "helix loop helix."  Fortunately we can do that too.  Just
change the link option to:

<blockquote class="example"><pre>
link         = http://www.google.com/search?q=$description
</pre></blockquote>

<p>

There are a large number of possible variables that you can use inside
link rules.  See the CONFIGURE_HOWTO document in the GBrowse
distribution for the full list.  You can also construct links using
Perl callbacks as described in the section on <a
href="#multiple_alignments">displaying ESTs</a>.  This gives you the
ability to generate any arbitrary URL.

<p>

If you want nothing to happen when the user clicks on a feature, just
set link to empty ("link = ").

<p>

The last thing we'll do is to change the behavior of the [Motif] track
so that:

<ol>
  <li>a new window pops up with the google search rather than
      replacing the contents of the current window
  <li>when the user mouses over a motif, a hints box will appear
      telling him that clicking there will initiate a google search
</ol>

<p>

These changes are easy:

<blockquote class="example"><pre>
[Motifs]
feature      = motif
glyph        = span
height       = 5
description  = 1
link         = http://www.google.com/search?q=$description
link_target  = _blank
title        = Search Google for $description.
key          = Example motifs
</pre></blockquote>

<p>

There's now a <code>link_target</code> option.  This contains the name
of a browser window in which to load the content when the user clicks
on the feature.  If there's no window of that name, the browser will
create a new window and give it the desired name.  Choose an ordinary
name like "Google" if you want the Google content to be loaded into
the same window each time, or choose "_blank" as we've done here in
order to pop up a new fresh window each time the user clicks.

<p>

The <code>title</code> option contains a bit of text that will be
displayed whenever the user hovers the mouse over the feature for a
second or two.  The same variable substitution rules apply, so when
the user mouses over feature "m06", a hints window will pop up that
says "Search Google for SUSHI repeat."  Give it a try!
</p>

<hr>

<h2><a name="enhancements">21. Enhancements</a></h2>

<p>
This section will show you how to add two nice user interface
enhancements to the volvox database.
</p>

<h3>Adding a "region" panel</h3>

<p> With larger genomes, you may want to add a "region panel" that is
intermediate in size between the overview panel and the detail
panel. The region panel can contain tracks of its own and is useful
for displaying features that are too numerous for the overview panel
and too large for the detail panel.</p>

<p>Open the volvox.conf configuration file and add the following line
to the [GENERAL] section. A good place is near the "max segment" and
"default segment" sections:</p>

<blockquote class="example"><pre>
# max and default segment sizes for detailed view
max segment     = 50000
default segment = 5000

# size of the "region panel"
region segment = 20000
</pre></blockquote>

<p>Now when you reload the volvox page, you will see an intermediate
panel labeled "region", as shown in Figure 22:</p>

<blockquote>
<img src="figures/enhancements1.gif"><br>
<i>Figure 24:</i> The "region" panel shows a region intermediate in
size between the overview and the detail panel.
</blockquote>

<p>You can declare region panel tracks in exactly the same way that you
declare overview tracks by declaring stanzas qualified by ":region"</p>

<blockquote class="example"><pre>
[TransChip:region]
feature        = tprofile
glyph          = xyplot
graph_type     = boxes
height         = 50
min_score      = 0
max_score      = 1000
bgcolor        = blue
scale          = right
key            = Profile
</pre></blockquote>

<p>Figure 25 shows what the region looks like with its "Profile" track
turned on.</p>

<blockquote>
<img src="figures/enhancements2.gif"><br>
<i>Figure 25:</i> You can add any number of tracks to the region
panel, just as you would for the overview panel.
</blockquote>

<h3>Grouping Tracks</h3>

<p>The bottom of the GBrowse window contains an expandable set of
checkboxes that allows the users to turn tracks on and off. By
default, the tracks are grouped into sections corresponding to tracks
belonging to the overview panel, those belonging to the region panel,
tracks created by external (third-party) annotations, and tracks
created by plugins. All other tracks are grouped together in a
catch-all section named "General."</p>

<p>You can easily define new track groups to make navigation
easier. To do so, just add a "category" option to each of the track
stanzas. This option defines the name of the category. Tracks that
belong to the same category will be grouped together, regardless of
the order in which the track definitions appear in the configuration
file. For example, we can place the [Motifs] and the [Translation]
tracks into a section named "Proteins" by modifying their stanzas
to look like this:</p>

<blockquote class="example"><pre>
[Motifs]
feature      = motif
glyph        = span
height       = 5
description  = 1
category     = Proteins
key          = Example motifs

[Translation]
glyph          = translation
global feature = 1
height         = 40
fgcolor        = purple
start_codons   = 0
stop_codons    = 1
category       = Proteins
translation  = 6frame
key          = 6-frame translation
</pre></blockquote>

<p>In this way we can create sections named "Alignments," "Examples,"
"Genes" and "Proteins" and assign the appropriate tracks to them. The
Tracks control section will look something like figure 24:</p>


<blockquote>
<img src="figures/enhancements3.gif"><br>
<i>Figure 26:</i> You can add any number of tracks to the region
panel, just as you would for the overview panel.
</blockquote>


<blockquote>
<i>The file <a href="conf_files/volvox_final.conf">volvox_final.conf</a>
contains the final configuration file with all the modifications we've
made during the course of this tutorial.  The data files <a
href="data_files/volvox_all.gff">volvox_all.gff</a> and <a
href="data_files/volvox_all.fa">volvox_all.fa</a> likewise contain the
entirety of the feature and DNA data.</i>
</blockquote>

<a name="cg"><h3>Display tracks within a group category as a table</h3>

<p>A further refinement to display track information within the category
 is a table display with headings for the rows and columns (see figure 27 for an example). 
 This layout is useful for displaying data that highlights the experimental design as in
 microarray or ChIP-on-Chip experiments.</p>

<blockquote>
<img src="figures/categorytable.png"><br>
<i>Figure 27:</i> An example of a category table containing a 9 track table,  
organized as 3 rows x 3 columns each with a heading. 
</blockquote>

<p>This was constructed using the option 'category tables' that is written at towards the top of 
the conf file. The first parameter refers to the category you wish to add the table to, the 
second is a space separated list of column headings, the third a 
space separated list of row headings.</p>

<blockquote class="example"><pre>
# category table configuration
category tables = 'ArrayExpts' 'strain-A strain-B strain-C' 'temperature anaerobic aerobic'
</blockquote>

It is then important that your stanzas within the category are in
column followed by row order (see example below and compare with
figure 27). So stanza 1 is column 1/row 1, stanza 2 is column 1/row 2,
stanza 3 is column 1/row 3, stanza 4 is column 2/row 1, stanza 5 is
column 2/row 2 etc. This means each cell in the table must have a
stanza. Any surplus tracks within that category will be ignored. For
example if there was a stanza 10, this would not be shown.  If there
are empty tracks they can be disabled using the 'disabled = 1' option
in the stanza. So to display the category table in figure 27 you would
use the following configuration.

<blockquote class="example"><pre>
[temp_strainA]
category       = ArrayExpts
feature        = temp_strainA_agg
glyph          = xyplot
bgcolor        = red
neg_color      = green
fgcolor        = black
graph_type     = boxes
height         = 80
min_score      = -2.0
max_score      = 2.0
scale          = both
key            = Temp strain A (1 expt)

[anaerobic_strainA]
category       = ArrayExpts
feature        = anaerobic_strainA_agg
glyph          = xyplot
bgcolor        = red
neg_color      = green
fgcolor        = black
graph_type     = boxes
height         = 80
min_score      = -2.0
max_score      = 2.0
scale          = both
key            = Anaerobic Strain A (0 expt)
disabled       = 1

[aerobic_strainA]
category       = ArrayExpts
feature        = aerobic_strainA_agg
glyph          = xyplot
bgcolor        = red
neg_color      = green
fgcolor        = black
graph_type     = boxes
height         = 80
min_score      = -2.0
max_score      = 2.0
scale          = both
key            = Aerobic Strain A (0 expt)
disabled       = 1


[temp_strainB]
category       = ArrayExpts
feature        = temp_strainB_agg
glyph          = xyplot
bgcolor        = red
neg_color      = green
fgcolor        = black
graph_type     = boxes
height         = 80
min_score      = -2.0
max_score      = 2.0
scale          = both
key            = Temp strain B (2 expts)

[anaerobic_strainB]
category       = ArrayExpts
feature        = anaerobic_strainB_agg
glyph          = xyplot
bgcolor        = red
neg_color      = green
fgcolor        = black
graph_type     = boxes
height         = 80
min_score      = -2.0
max_score      = 2.0
scale          = both
key            = Anaerobic Strain B (0 expt)
disabled       = 1

[aerobic_strainB]
category       = ArrayExpts
feature        = aerobic_strainB_agg
glyph          = xyplot
bgcolor        = red
neg_color      = green
fgcolor        = black
graph_type     = boxes
height         = 80
min_score      = -2.0
max_score      = 2.0
scale          = both
title          = blah
key            = Aerobic strain B (3 expts)

[temp_strainC]
category       = ArrayExpts
feature        = temp_strainC_agg
glyph          = xyplot
bgcolor        = red
neg_color      = green
fgcolor        = black
graph_type     = boxes
height         = 80
min_score      = -2.0
max_score      = 2.0
scale          = both
key            = Temp strain C (1 expt)

[anaerobic_strainC]
category       = ArrayExpts
feature        = anaerobic_strainC_agg
glyph          = xyplot
bgcolor        = red
neg_color      = green
fgcolor        = black
graph_type     = boxes
height         = 80
min_score      = -2.0
max_score      = 2.0
scale          = both
key            = Anaerobic strain C (3 expts)

[aerobic_strainC]
category       = ArrayExpts
feature        = aerobic_strainC_agg
glyph          = xyplot
bgcolor        = red
neg_color      = green
fgcolor        = black
graph_type     = boxes
height         = 80
min_score      = -2.0
max_score      = 2.0
scale          = both
key            = Aerobic strain C 3 (2 expts)
</blockquote>

<p>

If you need to have multiple category tables, simply use continuation
lines for the "category tables" option:

<blockquote class="example"><pre>
# category table configuration
category tables = 'ArrayExpts' 'strain-A strain-B strain-C' 'temperature anaerobic aerobic'
                  'CHiP-Chip'  'TFX1 ONE-CUT PHA4' '16-cell-stage 320-cell-stage adult'
</blockquote>

<hr>

<h2><a name="berkeleydb">22. Setting up a Berkeleydb database</a></h2>

<p>

The in-memory database is great for smaller data sets, and can handle
GFF files of up to about 20,000 features (more if you have lots of
memory).  For larger data sets, however, you'll want to use a database
management system.  GBrowse handles a number of DBMS through its
"database adaptor" system.  This section shows how to use the
Bio::DB::GFF berkeleydb adaptor that comes for free when you install
BioPerl; this will enable you to create databases of 10 million or
more features. The next section shows you how to install a MySQL
relational database that will support even larger data sets.  You may
skip these sections and move on to working with third-party
annotations if you do not wish to install a berkeleydb-based server at
this time.

<p>

The Berkeleydb database adaptor comes with BioPerl 1.51 or higher
(still under development at the time this tutorial was written). If
you have an older version of BioPerl, GBrowse will install the adaptor
for you. As its name implies, this adaptor uses the Berkeleydb
database system (http://www.sleepycat.com) to create indexed database
files from GFF feature files. The adaptor also requires the Perl
DB_File interface to Berkeleydb. If you are using a Linux or Mac OSX
system, you almost certainly have both Berkeleydb and DB_File already
installed. For Windows users of ActiveState Perl, you should confirm
that DB_File is installed by running the following command: </p>

<blockquote class="example">
<pre>
C:\&gt; perl -MDB_File -e 'print $DB_FILE::VERSION'
</pre>
</blockquote>

<p>
If this prints out a number, then you are golden. If you get an error,
you should reinstall DB_File by running the PPM tool:</p>

<blockquote class="example"><pre>
C:\&gt; ppm
PPM interactive shell (2.1) - type 'help' for available commands.
PPM> install DB_File
</pre></blockquote>

<p>It is an extremely simple task to convert an existing in-memory
database to use the Berkeleydb database. We will now convert the
Volvox example database to Berkeleydb.</p>

<p>Take the most recent version of the volvox.conf configuration file,
and edit the top few lines of the new file so that it looks like
this:</p>

<blockquote class="example"><pre>
[GENERAL]
description   = Volvox Berkeleydb Database
db_adaptor    = Bio::DB::GFF
db_args       = -adaptor berkeleydb
	        -dir     '/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases/volvox'
</pre></blockquote>

<p> We made just two changes. First, we changed the description of the
database to "Volvox Berkeleydb Database" to distinguish it from the
in-memory database. Second, we changed the value of the
<b>-adaptor</b> option from "memory" to "berkeleydb".
</p>

<p>Now reload the volvox page in your browser. There will be a slight
delay as the Berkeleydb adaptor constructs its indexes, and then the
page will reappear. You should now be able to browse and search the
database exactly as before. Depending on how fast the memory adaptor
was to begin with, you may not notice a speed improvement; however,
with large GFF files, the performance improvement will be very
marked.</p>

<p>If you look in the volvox database directory, you will see a series
of newly-created index files named "bdb_features.btree",
"bdb_features.data", etc. These are automatically created when needed
and updated whenever the underlying GFF or FASTA files are
changed.</p>

<p>If you get an "Internal Server Error" or similar message, check the
server error log file for messages that explain what went wrong. The
most common problem is that the volvox database directory is not
writeable by the web server user. As described earlier, this directory
must be "world writeable" in order to allow the web server to create
and maintain the databases</p>

<h3>Creating a Berkeleydb database using bp_load_gff.pl</h3>

<p>Although it is convenient to maintain the Berkeleydb indexes
automatically, this mechanism has a number of disadvantages. One
disadvantage is that this mechanism requires the database directory to
be world writeable (or at least writeable by the web user), which may
not be acceptable in some installations. Another disadvantage is that
the indexing may take a long time, up to 10 minutes for a GFF
databases containing a million lines. Some web servers will time out
during this process. For large databases, it is better to explicitly
create the database index files using the <i>bp_load_gff.pl</i>
program.</p>

<p><i>bp_load_gff.pl</i> is a BioPerl utility that is described in
more detail in <a href="#mysql">Setting up a MySQL database</a>. It
takes as its input a series of GFF and FASTA files and creates the
appropriate database files. To see how to use it, we will create a
fresh database directory. Go to the GBrowse database located at
<b>/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases</b> and create a new subdirectory called
"volvox_bdb:"</p>

<blockquote class="example"><pre>
 % cd /tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases
 % mkdir volvox_bdb
</pre></blockquote>

<p>

On Windows systems you can use the file manager to create this new
folder.

<p>

You do <b>not</b> have to make this directory world writeable, but it
should be readable and executable by the user that the web server runs
as. Now enter the tutorial data files directory
(/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/tutorial/data_files) and load the GFF and sequence
files using the following command:</p>

<blockquote class="example"><pre>
<b>% bp_load_gff.pl -c -a berkeleydb -d /tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases/volvox_bdb volvox_all.fa volvox_all.gff</b>
volvox_all.gff: loading...
738 records loaded
volvox_all.gff: 738 records loaded
Loading fasta file volvox_all.fa
volvox_all.fa: 7 records loaded
</pre></blockquote>

The arguments to <i>bp_load_gff.pl</i> are:

<table>
  <tr>
    <td><b>-a</b></td>
    <td>Use the berkeleydb database <b>a</b>daptor.
  </tr>
  <tr>
    <td><b>-c</b></td>
    <td><b>c</b>lear (initialize) the database
  </tr>
  <tr>
    <td><b>-d /tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases/volvox_bdb</b></td>
    <td>Load the data into the indicated <b>d</b>atabase directory.
  </tr>
  <tr>
    <td><b>volvox_all.fa volvox_all.gff</b></td>
    <td>The data files to load.
  </tr>
</table>

<p> If all goes well, this will create the index files in
<tt>/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases/volvox_bdb</tt>. If you look in that
directory now, you'll see a series of bdb_* index files.</p>

<p>
The last step is to modify the volvox.conf to point to this
directory. Open it in a text editor and modify the top part so that it
looks like this:
</p>

<blockquote class="example"><pre>
[GENERAL]
description   = Volvox Berkeleydb Database
db_adaptor    = Bio::DB::GFF
db_args       = -adaptor berkeleydb
	        -dsn    '/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases/volvox_bdb'
</pre></blockquote>

<p>The change here is to replace the <b>-dir</b> argument with
<b>-dsn</b> ("data source name").  This tells the Berkeleydb adaptor
that pre-made index files can be found in the indicated directory. It
will not attempt to update the index files automatically.</p>

<p>If you wish to update the indexes with new GFF or sequence data,
you should run the bp_load_gff.pl script again to update the
indexes. Using the <b>-c</b> flag will reinitialize the indexes from
scratch, erasing whatever was there before. Without this flag, the
provided GFF and/or sequence data will be incrementally added to the
indexes.</p>

<h2><a name="mysql">23. Setting up a MySQL database</a></h2>

<p>

The Bio::DB::GFF MySQL adaptor is an interface to the open source
MySQL database management system. Its performance is similar to that
of the Berkeleydb adaptor, but it has better provisions for error
recovery and is safe to use in environments where multiple users write
to the database simultaneously. In addition, the MySQL adaptor has
been tested much more extensively than the Berkeleydb adaptor and is
highly recommended for production environments. This section describes
how to set up GBrowse to use the MySQL adaptor. If you are not
interested in this, you may skip to the next section that describes
loading third-party annotations.

<p>

First you'll have to install MySQL.  Although it is installed by
default in most Linux systems, it will not be present on Windows or
Macintosh OSX systems.  Go to <a
href="http://www.mysql.com">www.mysql.com</a> and follow the
instructions to download and install the database.  Come back here
when this is done.

<p>

Next, you'll need to install the Perl interface to MySQL.  On a
Windows system using ActiveState Perl, use the ppm tool:

<blockquote><pre>
C:\Windows&gt; <b>ppm</b>
ppm&gt; <b>install DBD::mysql</b>
ppm&gt; <b>quit</b>
</pre></blockquote>

<p>

On a Unix, Linux or Mac OSX system, use the perl CPAN installer (this
may need to be done with root/superuser privileges):

<blockquote><pre>
% <b>perl -MCPAN -e shell</b>
cpan&gt; <b>install DBD::mysql</b>
cpan&gt; <b>quit</b>
</pre></blockquote>

<p>

Now you're ready to create the MySQL version of the volvox database.
First you'll set up a new empty database named "volvox."  Using the
<b>mysql</b> command-line tool, create the database, grant yourself
read/write privileges, and grant the "nobody" user read privileges:

<blockquote class="example"><pre>
% <b>mysql -uroot -p</b>
Enter password: *********

mysql> <b>create database volvox;</b>
Query OK, 1 row affected (0.04 sec)

mysql> <b>grant all privileges on volvox.* to lstein@localhost;</b>
Query OK, 0 rows affected (0.00 sec)

mysql> <b>grant select on volvox.* to nobody@localhost;</b>
Query OK, 0 rows affected (0.00 sec)

mysql> <b>quit</b>
Bye
</pre></blockquote>

<p>

Depending on how mysql was installed, you may not need to provide a
password, in which case just type "mysql -uroot" without the "-p"
argument.  When granting privileges to yourself, replace "lstein" with
your own login name.  If you are on a Windows system, you may be able
to skip this step entirely.

<p>

You'll now load the .gff and .fa files into this newly created
database.  There are actually two steps needed.  The first is to
"initialize" the database with all the data definitions needed to hold
genomic feature data, and the second is to actually load the data.
Fortunately, both these steps are handled by the same command-line
tool, <i>bp_load_gff.pl</i>, which is part of the BioPerl suite.

<p>

Copy the files <a href="data_files/volvox_all.gff">volvox_all.gff</a>
and <a href="data_files/volvox_all.fa">volvox_all.fa</a> to some
convenient place.  Then run the following command from the command
line:

<blockquote class="example"><pre>
% <b>bp_load_gff.pl -c -d volvox volvox_all.fa volvox_all.gff</b>
volvox_all.gff: loading...
volvox_all.gff: 738 records loaded
Loading fasta file volvox_all.fa
volvox_all.fa: 7 records loaded
</pre></blockquote>

<p>

The arguments to <i>bp_load_gff.pl</i> are:

<table>
  <tr>
    <td><b>-c</b></td>
    <td><b>c</b>lear (initialize) the database
  </tr>
  <tr>
    <td><b>-d volvox</b></td>
    <td>Load into the <b>d</b>atabase named volvox
  </tr>
  <tr>
    <td><b>volvox_all.fa volvox_all.gff</b></td>
    <td>The data files to load.
  </tr>
  
</table>

<p>

The MySQL database is all ready to go.  Now, in order to tell GBrowse
to start using the MySQL database rather than the in-memory database,
you need to make a small change to the volvox.conf configuration
file.  Find the few lines of the file and change them to look like
this:

<blockquote class="example"><pre>
[GENERAL]
description   = Volvox Example Database
db_adaptor    = Bio::DB::GFF
db_args       = -adaptor dbi::mysql
	        -dsn     volvox
                -user    nobody
                -pass
</pre></blockquote>

<p>

The <b>-adaptor</b> argument is telling GBrowse to use the
"dbi::mysql" database adaptor, which is the BioPerl interface to
MySQL databases.  The <b>-dsn</b> argument tells GBrowse to use the
<b>d</b>ata <b>s</b>ource <b>n</b>ame "volvox".

<p>

When you reload the web page, GBrowse will now be using MySQL.
Depending on the speed of your CPU and disk, you might notice that it
seems a bit snappier than the in-memory version.  See <a
href="../CONFIGURE_HOWTO.txt">CONFIGURE_HOWTO.txt</a> for more
information on configuring GBrowse to use relational databases.  Also
see the following perldoc manual pages:

<dl>
  <dt><b>perldoc Bio::DB::GFF::Adaptor::dbi::mysql</b>
  <dd>The MySQL adaptor.
      <p>
  <dt><b>perldoc Bio::DB::GFF::Adaptor::dbi::oracle</b>
  <dd>The Oracle adaptor.
      <p>
  <dt><b>perldoc Bio::DB::GFF::Adaptor::dbi::pg</b>
  <dd>The PostgreSQL adaptor.
      <p>
  <dt><b>perldoc Bio::DB::GFF::Adaptor::dbi::biofetch</b>
  <dd>An adaptor that will fetch data automatically from
      GenBank/EMBL and load it into a local MySQL database.
      <p>
  <dt><b>perldoc Bio::DB::GFF::Adaptor::memory</b>
  <dd>An adaptor for in-memory databases running off files.
      <p>
  <dt><b>perldoc Bio::DB::Das::Chado</b>
  <dd>An adaptor for PostgreSQL databases using the <i>Chado</i>
      schema (see the <a href="http://www.gmod.org/schema/">Chado
      home page</a>.)
      <p>
  <dt><b>perldoc Bio::DB::Das::BioSQL</b>
  <dd>An adaptor for PostgreSQL and MySQL databases using the
  <i>BioSQL</i> schema (see <a
  href="http://www.biosql.org">www.biosql.org</a>).
</dl>


<hr>

<h2><a name="third_party">24. Adding Third Party Features</a></h2>

<p>

It is often useful to have independent annotation data sets that can
be visualized together but updated separately.  For example, you may
be working on a genome that has a core set of stable annotations that
everyone shares, such as the set of protein-coding genes, and
independent sets of annotations that change frequently, such as
promoter predictions and experimental data.

<p>

GBrowse provides several mechanisms for making this type of modular
annotation possible.  You can:

<ol>
  <li>Upload one or more files of annotations temporarily, and view them in the
      context of the core annotations.  These annotations will be private to the
      user who uploads the annotations; others cannot see the data.
  <li>Put one or more GFF files in a web-accessible location, such as
      an FTP or Web site, and point GBrowse at it.  These annotations
      will be accessible to anyone who knows the correct URLs.
  <li>Point one GBrowse at another GBrowse.  All the tracks in the second
      instance of GBrowse will be available to the first GBrowse.
      This method uses the Distributed Annotation System (DAS) and can
      handle very large data sets.
</ol>

<p>

This section will lead you through the various ways to view third
party annotations on top of GBrowse. The examples are somewhat
contrived since we only have one computer to work with, and by
necessity both the main data and the third-party feature data will
have to reside on the same computer.  Don't be confused by this, and
keep in mind that in the real world, GBrowse will be running on one
computer, and the third-party annotation data will be loaded from
another network-accessible computer.

<p>

Instead of using the artificial volvox data, we will now use some real
genome annotations from the <i>C. elegans</i> genome project.  This is
a region around <i>C. elegans</i> cosmid C01F4.  The core data that
we'll be using is contained in the files <a
href="data_files/elegans_core.gff">elegans_core.gff</a>, and <a
href="data_files/elegans.fa">elegans.fa</a>.

<p>

Refer back to the <a href="#basics">beginning of the tutorial</a> now
and create a GBrowse database directory named "elegans_core".  Then
copy <a href="data_files/elegans_core.gff">elegans_core.gff</a>, and <a
href="data_files/elegans.fa">elegans.fa</a> into it.  The
configuration file to use is <a
href="conf_files/elegans_core.conf">elegans_core.conf</a>.  Place it in <a
href="file:////tmp/update-vendor-drop-jcQwDOU/FAKEROOT/conf/gbrowse.conf">/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/conf/gbrowse.conf/</a>.

<p>

Confirm that you can browse the database.  Figure 28 is a picture of
the entire data set with all core tracks turned on.

<blockquote>
<img src="figures/third_party1.gif"><br>
<i>Figure 28:</i> The core <i>C. elegans</i> dataset.
</blockquote>

<h3>Uploading an Annotation File</h3>

<p>

We will now add some third-party annotations to the display.  These
are contained in the files "elegans_acceptor.gff",
"elegans_expression.gff", "elefans_sts.gff", "elegans_deletion.gff", and
"elegans_repeats.gff":

<p>

<table border="1" width="100%">
  <TR>
    <td>
	<a href="data_files/elegans_acceptor.gff">elegans_acceptor.gff</a>
    </td>
    <td>
	Annotations of <i>C. elegans</i> spliced leader acceptor sites.
    </td>
  </TR>
    
  <TR>
    <td>
	<a href="data_files/elegans_expression.gff">elegans_expression.gff</a>
    </td>
    <td>
	Positions assayed for gene expression level in <i>C. elegans</i>
	microarrays.
    </td>
  </TR>
    
  <TR>
    <td>
	<a href="data_files/elegans_sts.gff">elegans_sts.gff</a>
    </td>
    <td>
	Primer pairs available for the region produced by the
	<i>C. elegans</i> ORFeome project.
    </td>
  </TR>

  <TR>
    <td>
	<a href="data_files/elegans_deletion.gff">elegans_deletion.gff</a>
    </td>
    <td>
	Deletion endpoints from a targeted gene knockout project.
    </td>
  </TR>
    
  <TR>
    <td>
	<a href="data_files/elegans_repeats.gff">elegans_repeats.gff</a>
    </td>
    <td>
	Complex repetitive elements found using the RepeatMasker program.
    </td>
  </TR>
    
</table>

<p>

We can load each of these files to private storage located on the
server using the file upload feature.  Copy these five files to your
home directory where you can find them easily.  Go to the section
marked <b>Upload your own annotations</b> and choose the "Browse..."
button.  Select one of the annotation files, and then press the
"Upload" button to upload the file to the server.  The annotations
contained in the file should now appear on the display.  If you now do
this for all five of the annotation files, you will eventually get a
display like that shown in Figure 29.

<p>

<blockquote>
<img src="figures/third_party2.gif"><br>
<i>Figure 29:</i> After uploading four annotation files.
</blockquote>

<p>

<blockquote>
NOTE: <i>This upload function works even if the gbrowse
you are uploading to is located on a remote server.  The uploaded
files are stored in a private directory on the server away from the
main data set.  Other users cannot see your data.</i>
</blockquote>

<p>

Although this display is functional, there is no difference between
the appearance of each of the tracks.  Fortunately, we can customize
the uploaded files quite easily.  Let us change the "elegans_sts.gff"
file so that the primer pairs use the "primers" glyph.  We can either
do this by deleting the uploaded file, making the appropriate
modification to our local version and then reuploading it, or by
editing the file in place.  We'll take the latter course.

<p>

Scroll to the bottom of the browser window, find the uploaded file
named "elegans_sts.gff", and choose "Edit File...".

<p>

<blockquote>
<img src="figures/third_party3.gif"><br>
<i>Figure 30:</i> The uploaded files can be edited in place by
clicking the "Edit File..." button.
</blockquote>

<p>

This will take you to a simple text editor window.  At the top of the
window, add the following configuration stanza:

<blockquote><pre>
# edited elegans_sts.gff file
[reagent]
glyph = primers
height = 6
key   = ORFeome project primer pairs
</pre></blockquote>

<p>

When you are done, press "Submit Changes..." and the display will be
updated to show the track with a more readable track name and the
primers glyph.

<p>

If you like, you can customize each of the files.  Here is a suggested
set of customizations:

<blockquote><pre>
# for the file elegans_repeats.gff
[repeat]
bgcolor = white
key = Complex repeats

# for the file elegans_acceptor.gff
[trans-splice_acceptor]
glyph = diamond
bgcolor = red
key = Trans-splice Acceptors

# for the file elegans_deletion.gff
[Deletion_allele]
glyph = span
key   = Gene knockouts

# for the file elegans_expression.gff
[Expression]
bgcolor = orange
height  = 4
key     = Microarray expression probe
</pre></blockquote>

<p>

With this combination of configurations, the display will now look
as shown in Figure 31:

<blockquote>
<img src="figures/third_party4.gif"><br>
<i>Figure 31:</i> After customizing the annotation files.
</blockquote>

<p>

<blockquote> NOTE: <i>Be aware of an important difference between the
track configuration of the uploaded files and of the main GBrowse
configuration files.  In GBrowse, the [STANZA] heading is the name of
the symbolic name of the track, and particular feature types are added
to the track using the feature= option.  In uploaded files, the
[STANZA] heading is the feature type itself.  This means that each
track can only contain one feature type.  However, any uploaded GFF
file can contain multiple feature types, and each feature type can
have its own configuration stanza.</i></blockquote>

<p>

<blockquote><i>The other important difference between the uploaded
file configuration and the GBrowse main configuration is that for
security reasons Perl subroutines are not allowed in the configuration
sections of uploaded files.  However links and link patterns are
allowed.</i></blockquote>

<p>

There is no particular reason that each of the annotation sets were
broken into separate files.  We could easily combine them into a
single GFF file just as you do for the core annotations.

<h3>Sharing an Annotation File</h3>

<p>

Once you have an uploaded annotation file set up the way you like it,
you might want to share it with others.  You can do this easily if you
have access to an anonymous FTP or web server (if you are reading this
tutorial, it is fair to assume that you do!)

<p>

To watch this in action, we will place one of the annotation files
onto the local web server and then load it from within the local
GBrowse.  This contrived example doesn't make much sense until you
realize that the same trick will work when the GBrowse server and the
web-accessible annotation file can be on separate machines halfway
across the world.

<p>

We will demonstrate using the elegans_sts.gff file.  Please use a
version that has been edited to place the [reagent] configuration
stanza at the top.  Then copy this file to the directory "/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs".
This will place it at the top of the Web server document tree, but
<b>outside</b> the location of GBrowse databases.  Check that the file
is correctly installed on your web server by fetching this URL: <a
href="elegans_sts.gff"
target="_new">http://localhost/elegans_sts.gff</a>.  If the file is
correctly installed on the Web server, you will see this:

<blockquote><pre>
[reagent]
glyph = primers
height = 6
key   = ORFeome project primer pairs

##gff-version 2
##date Tue Feb 24 06:39:41 2004
##sequence-region C01F4 1 40000
##source gbrowse GFFDumper plugin
##NOTE: Selected features dumped.
C01F4	Orfeome_project reagent 3319	17668	.  +  .  PCR_product mv_ZK783.1 ; Amplified 0
C01F4	Orfeome_project reagent 18584	20445	.  -  .  PCR_product mv_G_YK5686 ; Amplified 1
C01F4	Orfeome_project reagent 24509	25425	.  -  .  PCR_product mv_ZK783.3 ; Amplified 1
C01F4	Orfeome_project reagent 26525	33359	.  -  .  PCR_product mv_ZK783.4 ; Amplified 0
C01F4	Orfeome_project reagent 38660	49506	.  +  .  PCR_product mv_C18H2.1 ; Amplified 1
</pre></blockquote>

<p>

Now go back to your browser, and <b>delete</b> all the uploaded files.
(This is to prevent the list of tracks from getting too long!)  You
can do this by scrolling to the bottom of the browser window and
pressing "Delete File" for each of the annotation files that you
previously uploaded.  This should return you to the display of the
core gene models and EST alignments that we began with.

<p>

Now we'll reload the STS annotations by using their URL.  Scroll to
the bottom of the window, find the text field labeled "Enter Remote
Annotation URL", type in <b>http://localhost/elegans_sts.gff</b>, and
press "Update URLs."  The "ORFeome project primer pairs" track will
reappear.

<p>

In order to make this process even simpler, you can create a popup
menu containing the URLs of frequently-accessed remote annotation
files.  To make this more interesting, first copy the
elegans_expression.gff file to the "/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs" directory in the way
described earlier.  Now elegans_sts.gff and elegans_expression.gff
will be available as the URLs <a
href="/elegans_sts.gff">http://localhost/elegans_sts.gff</a>
and <a
href="/elegans_expression.gff">http://localhost/elegans_expression.gff</a>,
respectively.

<p>

Open up the GBrowse configuration file,
"/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/conf/gbrowse.conf/elegans_core.conf", and insert the following
lines right after the "plugins =" line:

<blockquote><pre>
# remote GFF files to make available for optional loading
remote sources = "ORFeome STSs"      http://localhost/elegans_sts.gff
	         "Expression probes" http://localhost/elegans_expression.gff

</pre></blockquote>

<p>

When you reload the web page, you will see a popup menu appear next to
the remote annotation URL textfield (Figure 32).  The menu will
contain options to load "ORFeome STSs" and "Expression probes", and
selecting a menu item will have exactly the same effect as typing in
the URL manually.

<blockquote>
<img src="figures/third_party5.gif"><br>
<i>Figure 32:</i> The preset remote annotation URL popup menu.
</blockquote>

The neat thing about all this is that it works across the
Internet. Send the URL of the annotation files to your colleagues
(being sure to replace "localhost" with the hostname of your web
server!) and they'll be able to load this URL into any GBrowse that
uses the same core annotations. You can also use this mechanism within
your laboratory or department to share annotation sets without having
to give everyone write access to the web server's /tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs directory.

<p>

To remove a URL from the list of loaded URLs, just delete it from its
text field and reload.

<hr>

<h2><a name="DAS">25. Using GBrowse as a DAS Server</a></h2>

<p>

The Distributed Annotation System protocol (DAS; <a
href="http://www.biodas.org">http://www.biodas.org</a>) is a system
for exchanging genomic annotations across the Internet.  It works
similarly to the idea of sharing the URLs of web-accessible GFF files,
except that it is designed to support large data sets.  When a client
application needs to fetch just a subset of the data, such as a small
piece of a chromosomal arm, the DAS protocol allows only the relevant
annotations to be retrieved, rather than the whole data set.

<p>

To take advantage of DAS functionality, you will have to install the
Perl Bio::Das module.  This is available from CPAN (the Comprehensive
Perl Archive Network (<a href="http://www.cpan.org"
target="_new">http://www.cpan.org</a>) or from the GMOD PPM
repository.  Unix users can install Bio::Das with this command:

<blockquote><pre>
% <b>perl -MCPAN -e 'install Bio::Das'</b>
</pre></blockquote>

<p>

Windows users can use the PPM tool:

<blockquote><pre>
C:\Windows&gt; <b>ppm</b>
ppm&gt; <b>install Bio::Das</b>
ppm&gt; <b>quit</b>
</pre></blockquote>

You may need to issue the command "rep add gmod
http://www.gmod.org/ggb/ppm" if PPM complains that it cannot find
Bio::Das.

<p>

When you installed GBrowse, you also installed a CGI script that
enables your web server to act as a DAS server. The CGI script is
named "/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/cgi-bin//das", and it runs off the same configuration files as
GBrowse itself.  Only a very small bit of extra configuration is
required to enable full DAS server functionality.  In this part of the
tutorial we will first turn on the DAS server, and then use it to
serve out annotations on the <i>C. elegans</i> database.

<p>

To start, open the elegans_core.conf configuration file and add the
following line to the configuration file.  It can go anywhere before
the start of the track definition stanzas, but it is probably a good
idea to place it towards the top between "plugins" and "default
features."

<blockquote>
<pre>
# DAS reference server
das mapmaster      = SELF
</pre>
</blockquote>

<p>

What this line is doing is to declare to the DAS system that our
server is authoritative for the coordinates on the current
<i>C. elegans</i> genome example.  This is appropriate if you are
starting out a genome for the first time.  If, however, you want to
annotate against an existing set of genome coordinates, you should
replace SELF with the URL of the DAS reference server that serves that
genome.  For example release <i>hg16</i> of the human genome at UCSC
corresponds to DAS URL http://genome.cse.ucsc.edu/cgi-bin/das.  A list
of reference servers for various model organisms can be found at <a
href="http://www.biodas.org">http://www.biodas.org</a>.

<p>

The next step is to go through the configured tracks and add a "das
category" to each of them.  DAS uses the idea of the "category" of a
feature in order to filter sets of features by their purpose.
Categories include:

<table border="1">
  <tr>
    <th>transcription</th>
    <td>features that have to do with
	RNA transcription</td>
  </tr>
  <tr>
    <th>translation</th>
    <td>features that have to do with
	protein translation and function</td>
  </tr>
  <tr>
    <th>variation</th>
    <td>mutations, deletions, polymorphisms</td>
  </tr>

  <tr>
    <th>structural</th>
    <td>contigs, clones, reads, PCR primers</td>
  </tr>

  <tr>
    <th>repeat</th>
    <td>repetitive elements</td>
  </tr>

  <tr>
    <th>experimental</th>
    <td>a catch-all for experimental data</td>
  </tr>

  <tr>
    <th>miscellaneous</th>
    <td>anything that doesn't fit in one fo the other categories</td>
  </tr>


</table>

<p>

Find the [Transcripts] stanza and modify it to to have a das category
of "transcription" as shown here:

<blockquote><pre>
[Transcripts]
feature      = processed_transcript
glyph        = processed_transcript
height       = 8
bgcolor      = blue
description  = 1
das category = transcription
key          = Protein-coding genes
</pre>
</blockquote>

Similarly, modify the [Alignments] track to have a das category of
"similarity."  You do not need to add a category to the DNA track, as
it is treated specially by das.  You're all done!  Be sure to save the
configuration file before you try the next step.

<p>

Using a web browser fetch the URL <a
href="/cgi-bin/das/dsn"
target="_new">http://localhost/cgi-bin/das/dsn</a>.  This will return
an XML document giving information about each of the data sources that
you have configured. 

<blockquote><pre>
&lt;?xml version="1.0" standalone="yes"?&gt;
&lt;!DOCTYPE DASDSN SYSTEM "http://www.biodas.org/dtd/dasdsn.dtd"&gt;
&lt;DASDSN&gt;
   &lt;DSN&gt;
      &lt;SOURCE id="elegans_core"&gt;elegans_core&lt;/SOURCE&gt;
      &lt;MAPMASTER&gt;http://localhost/cgi-bin/das/elegans_core&lt;/MAPMASTER&gt;
      &lt;DESCRIPTION&gt;C. elegans Core Annotations&lt;/DESCRIPTION&gt;
   &lt;/DSN&gt;
&lt;/DASDSN&gt;
</pre></blockquote>

This is showing that there is one configured DAS source, the
"elegans_core" data set.

<p>

Next test that the DAS "types" request is working.  This request
returns all the feature types that the database knows about.  Using a
web browser fetch the URL <a
href="/cgi-bin/das/elegans_core/types"
target="_new">http://localhost/cgi-bin/das/elegans_core/types</a>.
This should return another short document confirming that the
"processed_transcript" and "match:BLAT_EST_BEST" feature types are
available.

<p>

The final test that the DAS server is performing correctly is to
browse to the <a href="/cgi-bin/gbrowse/elegans_core"
target="_new">elegans_core</a> database and to <b>turn off</b> all the
tracks except for DNA/GC content.  This should give you an empty
details panel.  Now scroll down to the first empty URL entry field and
type in <b>http://localhost/cgi-bin/das/elegans_core</b> and press
"Update URLs."  The page should now reload and display the gene models
and the EST alignments.  However, the data is now not coming directly
from the local database, but from the database via the DAS protocol.

<h3>Combining Databases with DAS</h3>

We can now use DAS to integrate the core gene model and EST alignment
annotations with the STSs, expression data, trans-splice acceptors and
other third party annotations.  To do this, we will create a GBrowse
database that contains the third party annotations, but not the core
data.  This new database will be used as a DAS source.

<p>

Create a new database directory called elegans_extra in the
"/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/htdocs/gbrowse/databases" directory, and add to it a copy of the
file <a href="data_files/elegans_extra.gff">elegans_extra.gff</a>.
This GFF file is simply the result of concatenating together the
individual annotation files we looked at earlier (elegans_sts.gff,
etc), and removing the redundant comment lines from the top of the
file.  Now copy the configuration file <a
href="conf_files/elegans_extra.conf">elegans_extra.conf</a> into the
/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/conf/gbrowse.conf/ directory.  Have a look at this config file, and
note that it contains the appropriate "das mapmaster" and "das
category" configuration objects.

<p>

Once the config file is installed, confirm that you can browse the
extra annotations by fetching <a
href="/cgi-bin/gbrowse/elegans_extra"
target="_new">http://localhost/cgi-bin/gbrowse/elegans_extra</a>.

<p>

Now we're ready to layer the extra annotations onto the core
annotations using DAS.  Open up a browser window on the <a
href="/cgi-bin/gbrowse/elegans_core"
target="_new">http://localhost/cgi-bin/gbrowse/elegans_core</a>
database.  Delete any URLs that are already listed in the "Add remote
annotations" area, and add the URL
"http://localhost/cgi-bin/das/elegans_extra."  When you reload, the
core annotations will be shown on top, and the annotations from the
elegans_extra database will be shown in four tracks at the bottom of
the display.

<p>

The power of this feature is that we can use it across the Internet to
integrate databases that are independently maintained.  For example,
try adding the DAS URL
http://dev.wormbase.org/db/seq/das/elegans_even_more, and see what
appears.

<p>

By default, when you enter a DAS URL, the system will load all the
feature types that the DAS server makes available.  If this is not
desirable, you can limit the tracks by type and/or category.  To find
out what feature types a DAS server supports, retrieve a URL like the
following: <a
href="/cgi-bin/das/elegans_extra/types">http://localhost/cgi-bin/das/elegans_extra/types</a>.
This will provide a list of feature type names and their functional
categories.  From this we can see that the elegans_extra database
exports types of "repeat", "trans-splice_acceptor," "Deletion_allele,"
and "Expression."  Of course, we already knew this since we set the
database up ourselves!

<p>

Using this information, you can now limit the number of tracks
retrieved from the DAS server to just those that are of interest to
us. In the "Add remote annotations" text field, replace the current
DAS URL with this one:
http://localhost/cgi-bin/das/elegans_extra?type=repeat.  When you
reload, you will see only the repeat track and not the other three.

<p>

What if we want to see two of the four tracks?  We just add additional
type= sections, separated by semicolons.  To see both the "repeat" and
"Expression" tracks, we could request
http://localhost/cgi-bin/das/elegans_extra?type=repeat;type=Expression
(Figure 33).

<blockquote>
<img src="figures/DAS1.gif"><br> <i>Figure 33:</i>
The <i>C. elegans</i> core annotations database with the "repeat" and
"Expression" tracks  superimposed on it using DAS.
</blockquote>

<p>

To fetch features that match a particular category, we can add the
category= option to the URL.  For example, to fetch only features that
have to do with RNA transcription, you can request
http://localhost/cgi-bin/das/elegans_extra?category=transcription.

<p>

We can take advantage of this feature to add a menu of external DAS
annotations to the browser.  Open
"/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/conf/gbrowse.conf/elegans_core.conf" and insert the following
section right after the "plugins =" line:

<blockquote><pre>
# remote DAS data to make available for optional loading
remote sources =
   "DAS mRNA features"      http://localhost/cgi-bin/das/elegans_extra?category=transcription
   "DAS protein features"   http://localhost/cgi-bin/das/elegans_extra?category=translation
   "DAS repeat features"    http://localhost/cgi-bin/das/elegans_extra?category=repeat
   "DAS variation features" http://localhost/cgi-bin/das/elegans_extra?category=variation
   "DAS experimental features" http://localhost/cgi-bin/das/elegans_extra?category=experimental
</pre></blockquote>

When you reload, the page will now show a popup menu of pre-defined
DAS sources that users can choose.  The DAS sources can be local, as
shown here, or located on one or more remote web sites.

<h3>Exporting DAS Tracks to Ensembl and other Genome Browsers</h3>

<p>

GBrowse DAS tracks can be layered onto <a
href="http://www.ensembl.org">Ensembl</a> and other DAS-aware genome
browsers. There are a couple of things to bear in mind:

</p>

<ol>
  <li>Only the tracks explicitly labeled with "das category" will be
      exported.
  <li>The range of glyphs supported by Ensembl is more limited than GBrowse.
</ol>

<p>The last is a gotcha. The official list of DAS-recognized glyphs
can be found <a
href="http://www.biodas.org/documents/spec.html#glyphid">here</a>, but
gbrowse has a larger number of glyphs. Because of this, DAS-exported
features may not look on Ensembl the way they look on GBrowse. There
are three workarounds for this:

<dl>
  <dt>The <i>das flatten</i> option
  <dd>Set this option to flatten a multi-part feature, such as a gene,
  into a simpler "flat" structure that will display correctly on the
      Ensembl contig viewer. Also be sure to specify "grouping true"
      when you configure Ensembl for this DAS source.
  <dt>The <i>das glyph</i> option
  <dd>Set this option in an exported track stanza in order to force
  the glyph
      to a standard DAS glyph, such as "box". For example:
      <blockquote><pre>
      das glyph = box
      </pre></blockquote>
  <dt>The <i>das type</i> option
  <dd>Ensembl and possibly other browsers treat certain feature types
      specially. In particular, if a feature has a type of "gene" then
      Ensembl will display it with angled introns. Set <i>das type</i>
      in a track stanza to force the reported type to one of these
      special values. Example:
      <blockquote><pre>
      das type = gene
      </pre></blockquote>
</dl>


<h3>Running GBrowse off DAS Entirely</h3>

<p>

If you wish, you can even run GBrowse off a remote DAS server entirely
and keep no data locally (or just maintain private annotation tracks).
This works by replacing the Bio::DB::GFF database adaptor that we have
been using up to now with an adaptor named "Bio::Das". However,
because of a poorly characterized interaction between the Bio::Das
module and Perl 5.6, it is recommended that you use Perl 5.8.1 or
higher for this. Otherwise you may experience out of memory errors.

<p>To watch this in action, we will run GBrowse off the UCSC genome
browser, which exports its data in DAS format.

<p>

We will need a configuration file to do this.  DAS-based configuration
files are almost identical to the ones we have been using up to now
for local databases.  The main change is to replace the "db_adaptor" and
"db_args" options with ones appropriate for the DAS data source.  For
example, for the "hg16" human genome database maintained at UCSC, the
appropriate options will be:

<blockquote><pre>
[GENERAL]
description   = Human July 2003 Genome at UCSC
db_adaptor    = Bio::Das
db_args       = -source http://genome.cse.ucsc.edu/cgi-bin/das
	        -dsn    hg16
</pre></blockquote>

Conveniently enough, recent versions of the GBrowse distribution
include a utility called "make_das_conf.pl" that will build a basic
DAS browser configuration file for you.  This utility was installed
for you when you installed GBrowse. To run it, you will need to know
the base URL of the DAS server you're going to display.  For our
example, we'll use the UCSC DAS server at
http://genome.cse.ucsc.edu/cgi-bin/das.

<p>

This is a command-line utility.  To find out the databases served by
UCSC, type in the following command at the Unix or Windows command
line:

<blockquote><pre>
% <b>make_das_conf.pl http://genome.cse.ucsc.edu/cgi-bin/das</b>
The following DAS URLs are available at this server.  Please call the script again
using one of the following URLs:

http://genome.cse.ucsc.edu/cgi-bin/das/dm1
	Fruitfly Jan. 2003 Genome at UCSC

http://genome.cse.ucsc.edu/cgi-bin/das/hg13
	Human Nov. 2002 Genome at UCSC

http://genome.cse.ucsc.edu/cgi-bin/das/hg15
	Human April 2003 Genome at UCSC

http://genome.cse.ucsc.edu/cgi-bin/das/hg16
	Human July 2003 Genome at UCSC

http://genome.cse.ucsc.edu/cgi-bin/das/rn3
	Rat Jun 2003 Genome at UCSC
[... many many more ...]
</pre></blockquote>

We're looking for the hg16 release, so we reissue make_das_conf.pl
again using UCSC DAS server's URL with the hg16 release number
appended to the end:

<blockquote><pre>
% <b>make_das_conf.pl http://genome.cse.ucsc.edu/cgi-bin/das/hg16</b>
[GENERAL]
description   = Human July 2003 Genome at UCSC
db_adaptor    = Bio::Das
db_args       = -source http://genome.cse.ucsc.edu/cgi-bin/das
	        -dsn    hg16

# examples to show in the introduction
examples = 10 10_random 11 12 13 13_random 14 15 15_random
      16 17 17_random 18 18_random 19 19_random 1 1_random
      20 21 22 2 2_random 3 3_random 4 4_random 5 5_random
      6 6_random 7 7_random 8 8_random 9 9_random M
      Un_random X X_random Y

das mapmaster = http://genome.cse.ucsc.edu:80/cgi-bin/das/hg16

aggregators = ECgene{ECgene}
       affy10K{affy10K}
       affyGeno{affyGeno}
       affyRatio{affyRatio}
       affyTranscriptome{affyTranscriptome}
       affyU133{affyU133}
       affyU95{affyU95}
[...much much more...]
</pre>
</blockquote>

If you tried this at the command line, you saw a lot of text scroll up
your screen and disappear forever.  Run the command again, and this
time <b>redirect</b> its output into a new configuration file named
"ucsc_hg16.conf":

<blockquote><pre>
% <b>make_das_conf.pl http://genome.cse.ucsc.edu/cgi-bin/das/hg16 &gt;/tmp/update-vendor-drop-jcQwDOU/FAKEROOT/conf/gbrowse.conf/ucsc_hg16.conf</b>
</pre></blockquote>

That should be all you need to do, unless you are behind a firewall
that uses an HTTP proxy. In this case, you will need to edit the
"db_args" option in the generated configuration file to include a
-proxy option. This tells gbrowse to fetch the remote data using the
indicated proxy. For example:

<blockquote><pre>
[GENERAL]
description   = Human July 2003 Genome at UCSC
db_adaptor    = Bio::Das
db_args       = -source http://genome.cse.ucsc.edu/cgi-bin/das
	        -dsn    hg16
                -proxy  http://my.proxy.address
</pre></blockquote>

Try browsing the new data source by requesting <a
href="/cgi-bin/gbrowse/ucsc_hg16"
target="_new">http://localhost/cgi-bin/gbrowse/ucsc_hg16</a>, and you
should be able to browse through a rudimentary version of the Human
genome display.

<p>

Once you have a basic configuration file for a remote DAS source, you
can pretty it up by changing track styles, key names, and so
forth. Bear in mind that the make_das_conf.pl does its best to guess
about the right landmarks to use in the list of examples in the
instructions, which feature types should be made the defaults for
searching, and how to aggregate multi-part features together.  You
will almost certainly need to customize these options to meet your
needs.

<hr>

<h2><a name="conclusion">26. Conclusion</a></h2>

<p>

This is just a short introduction to the many things that you can do
with GBrowse.  Major features not discussed were:

<ul>
  <li>multi-language support
  <li>third-party feature loading
  <li>the ability to view GenBank, chado, and biosql feature databases
  <li>advanced callbacks
</ul>

All this information, and more can be found in CONFIGURE_HOWTO.txt,
and in the documentation for Bio::DB::GFF and Bio::Graphics in the
bioperl distribution.

<p>

Have fun!

<hr>
<address>Lincoln D. Stein, lstein@cshl.org<br>
<a href="/">Cold Spring Harbor Laboratory</a></address>
<!-- hhmts start -->
Last modified: Wed Mar 21 20:41:27 EDT 2007
<!-- hhmts end -->
</body> </html>
