package SGN::View::Toolbar;
use Moose;

extends 'Catalyst::View';

has 'toolbar_data' => (
    is       => 'rw',
    isa      => 'ArrayRef',
    required => 1,
    default  => \&default_toolbar_data
);

sub default_toolbar_data {
    [
        {
            name => 'search',
            link => '/search/direct_search.pl?search=loci',
            desc => 'Search our database.',
            menu => [

                {
                    name => 'Organisms',
                    link => '/search/organisms',
                    desc => 'search for organisms'
                },
                {
                    name => 'Genes',
                    link => '/search/locus',
                    desc => 'Gene search.'
                },
                {
                    name => 'Genome features',
                    link => '/search/features',
                    desc => 'Search for many types of annotations and sequences.'
                },
                {
                    name => 'Phenotypes',
                    link => '/search/stocks',
                    desc => 'Search for mutants, accessions, and traits.'
                },
                {
                    name => 'QTLs',
                    link => '/search/phenotypes/qtl',
                    desc => 'A web interface for uploading QTL raw data, on-the-fly QTL mapping and search interface for QTLs.'
                },
              #   {
#                     name => 'Genomic Selection',
#                     link => '/solgs/search/',
#                     desc => 'A tool for estimating breeding values using rrBlUP.'
#                 },
                {
                    name => 'Unigenes',
                    link => '/search/transcripts/unigene',
                    desc => 'Unigene search.'
                },
                {
                    name => "Gene Families",
                    link => "/search/family",
                    desc => "Unigene family search."
                },
                {
                    name => 'Markers',
                    link => '/search/markers',
                    desc => 'Marker search.'
                },
                {
                    name => 'BACs',
                    link => '/search/genomic/clones',
                    desc => 'BAC (Bacterial Artificial Chromosome) search.'
                },
                {
                    name => 'ESTs and libraries',
                    link => '/search/transcripts/est',
                    desc =>
'Find EST (Expressed Sequence Tag) libraries by keyword, e.g. library name, organism, tissue, development stage, or authors.'
                },
                {
                    name => 'Insitu database',
                    link => '/insitu/',
                    desc =>
'A database of in-situ images which can be updated by users, with image add, delete and annotation editing functions. The current images were generated by the floral genome project (FGP). For more information on the floral genome project, visit http://floralgenome.org/ and http://pgn.cornell.edu/.'
                },
                {
                    name => 'Images',
                    link => '/search/images',
                    desc => 'Search images contained in the SGN database.'
                },
                {
                    name => 'Expression',
                    link => '/search/expression',
                    desc => 'Search images contained in the SGN database.'
                },
                {
                    name => 'People',
                    link => '/search/people',
                    desc => 'Search database of researchers who use SGN.'
                },
                {
                    name => 'FAQ',
                    link => '/help/faq.pl',
                    desc => 'Frequently asked questions from users of SGN.',
                },
            ]
        },
        {
            name => 'maps',
            link => '/cview/',
            desc => 'View and compare chromosomes from various organisms.',
            menu => [
                {
                    name => 'Tomato-EXPEN 2000',
                    link =>
'/cview/map.pl?map_id=9&amp;show_offsets=1&amp;show_ruler=1',
                    desc =>
'A combined genetic and physical map, showing the association of BACs (Bacterial Artificial Chromosomes) and BAC-Contigs to this genetic map.',
                    style => 'color:#400'
                },
                {
                    name => 'Tomato-Kazusa/SolCAP',
                    link =>
'/cview/map.pl?map_id=31&amp;show_offsets=1&amp;show_ruler=1',
                    desc  => '',
                    style => 'color:#400'
                },
                {
                    name => 'Tomato QTLs',
                    link =>
'/cview/map.pl?map_id=34&amp;show_offsets=1&amp;show_ruler=1',
                    desc => '',
                    style => 'color:#400'

                },
                {
                    name => 'Tomato physical map',
                    link =>
'/cview/map.pl?map_id=p9&amp;show_offsets=1&amp;show_ruler=1',
                    desc =>
'This map is extracted from the S. lycopersicum LA925 x S. pennellii LA716 type F2.2000 genetic map.',
                    style => 'color:#400'

                },
                {
                    name => 'Tomato FPC contig map',
                    link =>
'/cview/map.pl?map_id=c9&amp;show_offsets=1&amp;show_ruler=1',
                    desc =>
'This map shows the positions of the Sanger HindIII/MboI FPC contigs relative to the the S. lycopersicum LA925 x S. pennellii LA716 type F2.2000 genetic map.',
                    style => 'color:#400'

                },

                {
                    name => 'Tomato IL map',
                    link =>
'/cview/map.pl?map_id=il6&amp;show_offsets=1&amp;show_ruler=1',
                    desc =>
'This map shows the locations of the introgressed fragments of S. pennellii in the S. lycopersicum genome, tied to the 1992 map.',
                    style => 'color:#400'
                },

                {
                    name => 'Tomato FISH map',
                    link => '/cview/map.pl?map_id=13',
                    desc =>
                      'Map of FISH (Fluorescent In Situ Hybridization) data.',
                    style => 'color:#400'
                },

                {
                    name => 'Tomato AGP map',
                    link => '/cview/map.pl?map_id=agp',
                    desc =>
'The Accessioned Golden Path (AGP) map for the tomato genome.',
                    style => 'color:#400'
                },
                {
                    name  => 'Pepper COSII map',
                    link  => '/cview/map.pl?map_version_id=58',
                    desc  => 'COSII map',
                    style => 'color:#530'
                },

                {
                    name => 'Pepper-AC99',
                    link => '/cview/map.pl?map_id=11',
                    desc =>
'This map is based on 100 individuals from an inter-specific F2 population of Capsicum. annuum cv. NuMex RNaky and Capsicum chinense var PI159234. 426 molecular markers, including 359 SSR markers 3 specific PCR markers and 68 RFLP markers were used to construct this linkage map, with a total length of 1304.8 cM.',
                    style => 'color:#530'
                },
                {
                    name => 'Pepper-FA03',
                    link => '/cview/map.pl?map_id=10',
                    desc =>
'This map is based on 100 individuals from an inter-specific F2 population of Capsicum. annuum cv. NuMex RNaky and Capsicum frutescens BG 2814-6. 728 molecular markers, including 489 SSR markers 195 AFLP, 8 specific PCR markers and 36 RFLP were used to construct this linkage map, with a total length of 1358.7 cM.',
                    style => 'color:#530'
                },
                {
                    name => 'Potato-TXB 1992',
                    link => '/cview/map.pl?map_id=3',
                    desc =>
'This is a map based on a S. tuberosum x S. berthaultii BC S. tuberosum population reported in Tanksley et al (1992).',
                    style => 'color:#220'
                },
                {
                    name  => 'Eggplant COSII map',
                    link  => '/cview/map.pl?map_version_id=63',
                    desc  => 'COSII map',
                    style => 'color:#202'
                },
                {
                    name => 'Eggplant-LXM 2002',
                    link => '/cview/map.pl?map_id=6',
                    desc =>
'This map is based on 58 F2 plants from an interspecific cross between S. linnaeanum (MM195) and S. melongena (MM738) and contains 233 RFLP markers.',
                    style => 'color:#202'
                },
                {
                    name  => 'Tobacco SSR 2007',
                    link  => '/cview/map.pl?map_id=15',
                    desc  => 'Nicotiana tabacum',
                    style => 'color:#333'
                },
                {
                    name => 'Arabidopsis COSII',
                    link =>
'/cview/map.pl?map_id=7&amp;show_offsets=1&amp;show_ruler=1',
                    desc  => 'Arabidopsis thaliana sequenced-based COSII map.',
                    style => 'color:#002800'
                },

                {
                    name => 'Comparative viewer...',
                    link =>
'/cview/view_chromosome.pl?show_offsets=1&amp;show_ruler=1&amp;show_offsets=1',
                    desc =>
                      'Compare chromosomes from different maps, side by side.',
                    style => 'font-style:italic;'
                },
                {
                    name  => 'See all available maps...',
                    link  => '/cview/index.pl',
                    desc  => 'All chromosome maps that are available on SGN.',
                    style => 'font-style:italic'
                }
            ]
        },
        {
            name => 'genomes',
            link => '/genomes/',
            desc => 'Read about sequencing projects.',
            menu => [
                {  name => 'Reference genomes' },
                {
                    name => 'Tomato',
                    link => '/organism/Solanum_lycopersicum/genome',
                    desc =>
'The tomato genome has been sequenced by an international consortium.',
                },
               {
                    name => '<i>Solanum pennellii</i>',
                    link => '/organism/Solanum_pennellii/genome',
                    desc => 'Wild tomato genome',
                },


                {
                    name => 'Potato',
                    link => '/organism/Solanum_tuberosum/genome',
                    desc => '',
                },
               {
                    name => 'Pepper',
                    link => '/organism/Capsicum_annuum/genome',
                    desc => '',
                },
		{ name => 'Draft genomes' },
                {
                    name => '<i>Solanum pimpinellifolium</i>',
                    link => '/organism/Solanum_pimpinellifolium/genome',
                    desc => 'Draft genome sequence of tomato\'s closest wild relative.',
                },
                {
                    name => '<i>Nicotiana benthamiana</i>',
                    link => '/organism/Nicotiana_benthamiana/genome',
                    desc => 'Draft sequence for Nicotiana benthamiana, sequenced by labs at the Boyce Thompson Institute',
                },
                {
                    name => '<i>Nicotiana tabacum</i> &amp; related',
                    link => '/organism/Nicotiana_tabacum/genome',
                    desc => 'Draft sequence for Nicotiana tabacum accessions K326, BX, and TN90. Parental genomes <i>Nicotiana sylvestris</i> and <i>Nicotiana tomentosiformis</i>.',
                },

		{ name => 'Resequencing projects' },
                {
                    name => '<i>Solanum lycopersicum</i> inbreds',
                    link => '/organism/Solanum_lycopersicum/inbred_genomes',
                    desc => 'Resequencing projects of tomato inbred lines'
                },
		 {
                    name => '150 Tomato Genome Resequencing Project',
                    link => '/organism/Solanum_lycopersicum/tomato_150',
                    desc => 'Resequencing project by the University of Wageningen'
                },
		{
                    name => 'BGI Tomato 360 genomes',
                    link => '/organism/Solanum_lycopersicum/tomato_360',
                    desc => 'Resequencing project of 360 tomato varieties by BGI'
                },

                {   name => 'Projects', type => 'title'  },

                {
                    name => 'Solanaceae project (SOL)',
                    link => '/solanaceae-project/index.pl',
                    desc =>
'Over the coming decade the International Solanaceae Genome Project (SOL) will create a coordianted network of knowledge about the Solanaceae family aimed at answering two of the most important questions about life and agriculture: How can a common set of genes/proteins give rise to such a wide range of morphologically and ecologically distinct organisms that occupy our planet? How can a deeper understanding of the genetic basis of diversity be harnessed to better meet the needs of society in an environmentally-friendly way? On this page, you will find more information about the strategy that will be used to answer these questions.'
                },

                {
                    name => 'SOL-100 Project',
                    link => '/organism/sol100/view',
                    desc => 'The SOL community aims to create a common Solanaceae-based genomic framework that includes sequences and phenotypes of 100 genomes encompassing the phylogenetic diversity of the group.',
                },

                {
                    name => 'International Tomato Sequencing Project',
                    link => '/organism/Solanum_lycopersicum/genome',
                    desc =>
'The gene-rich euchromatic portion of the tomato genome is being sequenced by an international consortium. This page gives more information about the progress of the sequencing effort.'
                },
                 {
                    name => 'U.S. tomato sequencing project',
                    link => '/about/us_tomato_sequencing.pl',
                    desc => 'An overview of the US Tomato Sequencing Project'
                },
                {
                    name => '<i>Nicotiana benthamiana</i> BTI site',
                    link => 'http://bti.cornell.edu/research/projects/nicotiana-benthamiana',
                    desc => 'BTI project for the Nicotiana benthamiana Draft sequence',
                },

            ]
        },
        {
            name => 'tools',
            link => '/tools/',
            desc => 'Tools for biologists.',
            menu => [

                #Sequence Analysis
                { name => 'Sequence Analysis' },
                {
                    name => 'BLAST',
                    link => '/tools/blast/',
                    desc =>
'Compare a given sequence to known sequences using the Basic Local Alignment Search Tool, NCBI BLAST v. 2.2.9 [May-01-2004]'
                },
                {
                    name => 'VIGS Tool',
                    link => 'http://vigs.solgenomics.net/',
                    desc =>
'Determine which genes a probe will silence using VIGS.',
		    target => '_blank',

                },

                {
                    name => 'Alignment Analyzer',
                    link => '/tools/align_viewer/',
                    desc =>
'Calculate alignments using Muscle v3.6, or view existing alignments using our navigator.'
                },
                {
                    name => 'Tree Browser',
                    link => '/tools/tree_browser/',
                    desc =>
'Visualize and manipulate phylogenetic trees from a newick-formatted string.'
                },
#                 {
#                     name => 'Intron Finder',
#                     link => '/tools/intron_detection/find_introns.pl',
#                     desc =>
# 'The SGN Intron Finder works by doing a BLAST search for <em>Arabidopsis thaliana</em> proteins that are similar to the translated protein sequence of the DNA input.'
#                 },

                #Mapping
                { name => 'Mapping' },
                {
                    name => 'Genome Browser (Jbrowse)',
                    link => '/jbrowse_solgenomics/',
                    desc => 'browse genomic annotations using the fast Jbrowse browser',
                },

                # {
                #     name => 'Genome Browser (Gbrowse)',
                #     link => '/gbrowse/bin/gbrowse/',
                #     desc => 'browse genomic annotations using GBrowse',
                # },
		# {
		#     name => 'Comparative Genome Browser',
		#     link => 'http://solgenomics.net/gbrowse2/bin/gbrowse_syn/sol2/?search_src=pot;name=ST3.10ch01:1..90000000',
		#     desc => 'Compare genomes using GBrowse_syn',
		# },
                {
                    name => 'Comparative Map Viewer',
                    link =>
'/cview/view_chromosome.pl?show_offsets=1&amp;show_ruler=1&amp;show_offsets=1',
                    desc =>
                      'Compare chromosomes from different maps, side by side.',
                },
                {
                    name => 'CAPS Designer',
                    link => '/tools/caps_designer/caps_input.pl',
                    desc =>
'Designs CAPS (Cleaved Amplified Polymorphic Sequence) assays for up to twelve sequences. Two types of nucleotide inputs are accepted: fasta sequences and clustal aligment. It generates a list of polymorphic enzymes that cut the sequences into different length products.'
                },
#                 {
#                     name => 'Seed BAC Finder',
#                     link => '/tools/seedbac/',
#                     desc =>
# 'Lists all anchored BACs for a given chromosome to help identify seed BACs, or suggests a seed BAC given a marker name. '
#                 },
		{
                    name => 'solQTL: QTL Mapping',
                    link => '/search/phenotypes/qtl',
                    desc =>
'A web interface for uploading raw QTL data, on-the-fly QTL mapping, linking predicted QTLs to relevant genetic and genomic data in SGN, and search interface for QTLs.'
                },
 #                {
#                     name => 'solGS: Genomic Selection',
#                     link => '/solgs/search',
#                     desc =>
#                         'Builds genomic selection models and predicts genomic estimated breeding values. Prediction is done using Ridge Regression - Best Linear Unbiased Prediction (RR-BLUP).'
#                 },
                #Molecular Biology
                { name => "Molecular Biology" },

#                {
#                    name => 'Signal Peptide Finder',
#                    link => '/tools/sigpep_finder/input.pl',
#                    desc =>
#'This method of searching for signal sequences is designed to complement SignalP, and has similar success rates.'
 #               },
                {
                    name => 'In Silico PCR',
                    link => '/tools/insilicopcr',
                    desc => 'In Silico PCR tool based on BLAST'
                },
		{
		    name => 'Tomato Expression Atlas (TEA)',
		    link =>'http://tea.solgenomics.net/',
		    desc => 'The Tomato Expression Atlas',
		},
		{
		    name => 'Tomato Expression Database (TED)',
		    link => '/ted',
		    desc => 'The Tomato Expression Database',
		},
                #Systems Biology
                { name => "Systems Biology" },
                {
                    name => 'SolCyc Biochemical Pathways',
                    link => '/tools/solcyc/',
                    desc => 'An interactive map of metabolic pathways.'
                },

                {
                    name => 'Coffee Interactomic Data',
                    link => '/tools/interactomics',
                    desc => 'Coffee Interactome Data with Cytoscape Web Start viewer.'
                },
                {
                    name => 'SGN Ontology Browser',
                    link => '/tools/onto/',
                    desc => 'AJAX ontology browser.'
                },

		{ name => "Breeder Tools" },
		{
		    name => 'Breeder Home',
		    link => '/breeders/home',
		    desc => 'Manage a breeding program',
		},
                #Bulk Query
                { name => 'Bulk Query' },

                {
                    name => 'FTP Site',
                    link => '/tools/bulk?mode=ftp',
                    desc =>
'FTP (File Transfer Protocol) is more efficient than HTTP (HyperText Transfer Protocol) for transferring large files. Many of our larger archive files are accessible here.'
                },
                {
                    name => 'Download Gene Sequences',
                    link => '/tools/bulk?mode=genes',
                    desc => 'Download nucleotide or protein sequences for a list of genes'
                },
                {
                    name => 'Clones, Arrays, Unigenes and BACs',
                    link => '/tools/bulk',
                    desc =>
'Rather than sifting through SGN page by page, here you can download large quantities of information in a single step.'
                },
                {
                    name => 'Unigene Converter',
                    link => '/tools/bulk?mode=converter',
                    desc => 'get the last version of unigenes and tomato gene ids from a list of unigenes.'
                },
#                 {
#                     name => 'ID Converter (SGN <=> TIGR)',
#                     link => '/tools/convert/input.pl',
#                     desc =>
# 'The Institute for Genomic Research and SGN maintain independent unigene databases, entries in which tend to have common member ESTs (Expressed Sequence Tags), although they tend not to correspond completely. This tool uses common members to convert back and forth between the two identifier sets.'
#                 },

                #Other
                # { name => "Other" },
                # {
                #     name => 'Primer 3',
                #     link => '/primer3/',
                #     desc => 'Pick primers for PCR reactions'
                # },
            ]
        }
    ];
}

=head2 headings

    my @headings=$tb->headings();

Returns a list of the names of the main toolbar buttons
('search','maps','genomes','tools'). These can then be used to
access other parts of this object.

=cut

sub headings {
    map { $_->{name} } @{ shift->toolbar_data };
}

=head2 heading_link

    print '<a href="'.$_->heading_link($heading).qq|">$_</a>|
        for @headings;

Takes a heading (which can be gotten from the headings
function). Returns the stuff that is supposed to go in the href
attribute for this heading.

=cut

sub heading_link {
    my $self = shift;
    my ($heading) = @_;
    for my $heading_hash ( @{ $self->toolbar_data } ) {
        if ( $heading_hash->{name} eq $heading ) {
            return $heading_hash->{link};
        }
    }
}

=head2 heading_desc

    for my $heading(@headings)
    {
        print"$heading description: ".$tb->heading_desc($heading);
    }

Takes a heading (which can be gotten from the headings function). Returns a description of what the linked-to page is supposed to do.

=cut

sub heading_desc {
    my $self = shift;
    my ($heading) = @_;
    for my $heading_hash ( @{ $self->toolbar_data } ) {
        if ( $heading_hash->{name} eq $heading ) {
            return $heading_hash->{desc};
        }
    }
}

=head2 menu_options

    for my $heading(@headings)
    {
        my @menu_options=$tb->menu_options($heading);
    }

Takes a heading (which can be gotten from the headings function). Returns a list of the names of the options below that heading ('Genes','Unigenes','Markers','BACs','Annotations','ESTs and libraries','People'). Like the list returned by the headings function, these can then be used to access other information from this object.

=cut

sub menu_options {
    my $self = shift;
    my ($heading) = @_;
    for my $heading_hash ( @{ $self->toolbar_data } ) {
        if ( $heading_hash->{name} eq $heading ) {

#            the toolbar links don't work if you add tooltips to the toolbar. fix?
#            my @menu_options=map {CXGN::Page::FormattingHelpers::tooltipped_text($_->{name},$_->{desc},'toolbar_help')} @{$heading_hash->{menu}};
            my @menu_options = map { $_->{name} } @{ $heading_hash->{menu} };
            return @menu_options;
        }
    }
}

=head2 option_link

    for my $heading(@headings)
    {
        my @menu_options=$tb->menu_options($heading);
        for my $option(@menu_options)
        {
            print"<a href=\"".$tb->option_link($heading,$option)."\">$option</a>";
        }
    }

Takes a heading and an option. Returns the stuff that goes in the href for that option.

=cut

sub option_link {
    my $self = shift;
    my ( $heading, $option ) = @_;

    #print STDERR "option_link received: heading '$heading' option '$option'\n";
    for my $heading_hash ( @{ $self->toolbar_data } ) {
        if ( $heading_hash->{name} eq $heading ) {
            for my $menu_option_hash ( @{ $heading_hash->{menu} } ) {
                if ( $menu_option_hash->{name} eq $option ) {
                    my $option_link = $menu_option_hash->{link};
                    if ($option_link) { return $option_link }
                }
            }
        }
    }

    #die"Option link not found for heading '$heading' option '$option'";
}

sub option_target {
    my $self = shift;
    my ( $heading, $option ) = @_;

    #print STDERR "option_link received: heading '$heading' option '$option'\n";
    for my $heading_hash ( @{ $self->toolbar_data } ) {
        if ( $heading_hash->{name} eq $heading ) {
            for my $menu_option_hash ( @{ $heading_hash->{menu} } ) {
                if ( $menu_option_hash->{name} eq $option ) {
                    my $option_target = $menu_option_hash->{target};
                    if ($option_target) { return ' target=$option_target ' }
                }
            }
        }
    }

}

=head2 option_style

    for my $heading(@headings)
    {
        my @menu_options=$tb->menu_options($heading);
        for my $option(@menu_options)
        {
            my $style = $tb->option_style($heading,$option);            		print "<span style='$style'>";
				#print option item
			  print "</span>";
        }
    }

Takes a heading and an option. Returns the stuff that goes in the href for that option. Override me.


=cut

sub option_style {
    my $self = shift;
    my ( $heading, $option ) = @_;

    #print STDERR "option_link received: heading '$heading' option '$option'\n";
    for my $heading_hash ( @{ $self->toolbar_data } ) {
        if ( $heading_hash->{name} eq $heading ) {
            for my $menu_option_hash ( @{ $heading_hash->{menu} } ) {
                if ( $menu_option_hash->{name} eq $option ) {
                    my $option_style = $menu_option_hash->{style};
                    if ($option_style) { return $option_style }
                }
            }
        }
    }

    #die"Option link not found for heading '$heading' option '$option'";
}

=head2 option_desc

    for my $heading(@headings)
    {
        my @menu_options=$tb->menu_options($heading);
        for my $option(@menu_options)
        {
            print"$option description: ".$tb->option_desc($heading,$option);
        }
    }

Takes a heading and an option. Returns the stuff that goes in the href for that option.

=cut

sub option_desc {
    my $self = shift;
    my ( $heading, $option ) = @_;
    for my $heading_hash ( @{ $self->toolbar_data } ) {
        if ( $heading_hash->{name} eq $heading ) {
            for my $menu_option_hash ( @{ $heading_hash->{menu} } ) {
                if ( $menu_option_hash->{name} eq $option ) {
                    return $menu_option_hash->{desc};
                }
            }
        }
    }
}

sub as_html {
    my $self     = shift;
    my @headings = $self->headings();

    #decorate the headings array with its menu alignments
    @headings = map [ 'center', $_ ], @headings;
    if ( @headings > 1 ) {
        $headings[0]->[0] = 'left';

        #      $headings[-1]->[0] = 'right';
    }

    my $index        = 0;
    my $toolbar_html = <<EOH
<table class="toolbar" summary="" cellspacing="0" cellpadding="0">
<tr class="toolbar_content">
  <td class="toolbar_l"><img src="/documents/img/toolbar_l.gif" alt="" /></td>
EOH
      . join(
        '',
        map {
            my ( $align, $heading ) = @$_;
            my $heading_link = $self->heading_link($heading);
            my $popup_menu   = $self->render_menu($heading);
            <<EOH
  <td class="toolbar_content" onmouseover="CXGN.Page.Toolbar.showmenu('$heading')" onmouseout="CXGN.Page.Toolbar.hidemenu()" style="text-align: $align">
    <a class="toolbar_menuname" href="$heading_link">$heading</a>
    <div style="position: relative; text-align: left">
$popup_menu
    </div>
  </td>
EOH
          } @headings
      )
      . <<EOH

  <td class="toolbar_search toolbar_content">
    <form name="quicksearch" class="quicksearch" action="/search/quick_search.pl">
      <input class="quicksearch field" type="text" name="term" size="14" />
      <input class="quicksearch imgbutton" type="image" src="/documents/img/sol_search_button.gif" value="sol search" />
    </form>
  </td>
  <td class="toolbar_r" height="1">
    <img src="/documents/img/toolbar_r.gif" alt="" />
  </td>
</tr>
</table>
<script language="JavaScript" type="text/javascript">
EOH
      . join( '', map <<EOH, @headings )
  CXGN.Page.Toolbar.addmenu('$_->[1]');
EOH
      . <<EOH
</script>
EOH
}

#return HTML for the popup menu for the  menu heading
sub render_menu {
    my ( $self, $heading ) = @_;

    my @menu_options = $self->menu_options($heading);

    return <<EOH
    <table cellspacing="0" cellpadding="0" id="$heading" class="toolbar_popmenu">
EOH
      . join( '',
              map {
                  my $menu_option = $_;

                  #menu options for each heading, these will be invisible unless hovered over
                  my $option_link = $self->option_link( $heading, $menu_option );
		  my $option_target = $self->option_target( $heading, $menu_option );
                  $option_link &&=
                      qq|<a class="toolbar_item" $option_target href="$option_link">$menu_option</a>|;
                  $option_link ||= qq|<span class="toolbar_item">$menu_option</span>|;
                  my $style = $self->option_style( $heading, $menu_option );
                  <<EOH
       <tr><td class="toolbar_item">$option_link</td></tr>
EOH
              } @menu_options
           )
      . <<EOH
    </table>
EOH
}

sub index_page {
    my $self         = shift;
    my ($heading)    = @_;
    my $content      = '';
    my $disp_heading = ucfirst($heading);
    for my $option ( $self->menu_options($heading) ) {
        my $link = $self->option_link( $heading, $option );
        my $desc = $self->option_desc( $heading, $option );
        if ($link) {
            $content .=
qq|<div style="margin-left:10px"><a href="$link">$option</a> - $desc<br /><br /></div>|;
        }
        else {
            $content .=
"<br /><span style='font-size:1.00em; font-weight:bold'>$option</span> $desc";
            $content .= ": $desc" if $desc;
            $content .= "<br /><br />";
        }
    }
    return <<END_HEREDOC;
<center>
<table class="boxbgcolor2" width="100%" summary="">
<tr>
<td width="15%">&nbsp;</td>
<td width="70%" class="left">
<div class="boxcontent">
<div class="subheading" style='font-size:1.1em'>
$disp_heading<br />
</div>
<div class="boxsubcontent">
$content
</div>
</div>
</td>
<td width="15%">&nbsp;</td>
</tr>
</table>
</center>
END_HEREDOC
}

1;
