<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>GBROWSE_IMG</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:root@localhost" />
</head>

<body style="background-color: white">

<p><a name="__index__"></a></p>
<!-- INDEX BEGIN -->

<ul>

	<li><a href="#gbrowse_img">GBROWSE_IMG</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<ul>

		<li><a href="#cgi_arguments">CGI arguments</a></li>
	</ul>

	<li><a href="#author">AUTHOR</a></li>
</ul>
<!-- INDEX END -->

<hr />
<p>
</p>
<h1><a name="gbrowse_img">GBROWSE_IMG</a></h1>
<p>The gbrowse_img script is a CGI script to generate genome images via
the Generic Genome Browser.</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<pre>
 * &lt;img src=&quot;<a href="http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=mec-3;width=400">http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=mec-3;width=400</a>&quot;&gt;</pre>
<pre>
     Will generate a picture of the mec-3 region of the &quot;wormbase&quot;
     data source.</pre>
<pre>
 * &lt;iframe src=&quot;<a href="http://localhost/cgi-bin/gbrowse_img/wormbase?name=B0001;embed=1&quot;width=&quot;100%&quot">http://localhost/cgi-bin/gbrowse_img/wormbase?name=B0001;embed=1&quot;width=&quot;100%&quot</a>; height=&quot;250&quot;&gt;
     &lt;img src=&quot;<a href="http://localhost/cgi-bin/gbrowse_img/wormbase?name=B0001&quot;/&gt">http://localhost/cgi-bin/gbrowse_img/wormbase?name=B0001&quot;/&gt</a>;
   &lt;/iframe&gt;</pre>
<pre>
 * &lt;a href=&quot;<a href="http://www.wormbase.org/db/seq/gbrowse_img?list=sources">http://www.wormbase.org/db/seq/gbrowse_img?list=sources</a>&quot;&gt;list&lt;/a&gt;</pre>
<pre>
     Will return this document: 
        ## Sources
        fly
        yeast
        wormbase</pre>
<pre>
 * &lt;a href=&quot;<a href="http://www.wormbase.org/db/seq/gbrowse_img/wormbase?list=types">http://www.wormbase.org/db/seq/gbrowse_img/wormbase?list=types</a>&quot;&gt;types&lt;/a&gt;</pre>
<pre>
     Will return this document:</pre>
<pre>
     ## Feature types for source wormbase
     tRNA       tRNAs 
     NG         Named Genes                default 
     CG         Curated genes              default 
     PG         Predicted genes 
     WABA       Briggsae alignments (WABA) 
     ESTB       ESTs aligned by BLAT (best) 
     ESTO       ESTs aligned by BLAT (other) 
     mRNAB      mRNAs aligned by BLAT (best) 
     mRNAO      mRNAs aligned by BLAT (other) 
     RNAi       RNAi experiments 
     EXPR       Expression chip profiles 
     WTP        Worm Transcriptome Project genes 
     SNP        SNPs 
     TcI        Transposon Insertions</pre>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This CGI script is an interface to the Generic Genome Browser for the
purpose of retrieving dynamic images of a region of the genome. It can
be used as the destination of an &lt;img&gt; tag like this:</p>
<pre>
 &lt;img src=&quot;<a href="http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=III:1..1000">http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=III:1..1000</a>&quot;&gt;</pre>
<p>The script can also be used to superimpose one or more external
features onto the display, for example for the purpose of displaying
BLAST hits, an STS or a knockout in the context of the genome.</p>
<p>
</p>
<h2><a name="cgi_arguments">CGI arguments</a></h2>
<p>The script recognizes the following CGI arguments, which can be passed
either as GET or POST argument=value pairs. Argument pairs must be
separated by semicolons (preferred) or by ampersands. Many of the
options have one-letter aliases that can be used to reduce URL
lengths.</p>
<pre>
   Argument     Alias   Description
   --------     -----   -----------
   name         q       genomic landmark or range
   type         t       tracks to include in image
   width        w       desired width of image
   options      o       list of track options (compact, labeled, etc)
   abs          b       display position in absolute coordinates
   add          a       added feature(s) to superimpose on the image
   style        s       stylesheet for additional features
   keystyle     k       where to place the image key
   overview             force an overview-style display
   flip         f       flip image left to right
   embed                generate full HTML for image and imagemap 
                        for use in an embedded frame
   format               format for the image (use &quot;SVG&quot; for 
                        scaleable vector graphics)
   list                 get certain types of configuration information
   source               database name</pre>
<p>The arguments are explained in more detail here</p>
<dl>
<dt><strong><a name="item_name">name (Alias: q)</a></strong>

<dd>
<p>This argument specifies the region of the genome to be
displayed. Several forms are recognized:</p>
</dd>
<dd>
<pre>
   + name=Landmark</pre>
</dd>
<dd>
<pre>
     Display the landmark named &quot;Landmark&quot;. Valid landmark names
     include chromosomes, contigs, clones, STSs, predicted genes, and
     any other landmark that the administrator has designated. Be
     careful when fetching large landmarks such as whole chromosomes!</pre>
</dd>
<dd>
<pre>
   + name=Landmark:start..end</pre>
</dd>
<dd>
<pre>
     Display the region between start and end relative to &quot;Landmark&quot;.</pre>
</dd>
<dd>
<pre>
   + name=Class:Landmark</pre>
</dd>
<dd>
<pre>
     Display &quot;Landmark&quot;, restricting to a particular class, such as
     &quot;PCR_Product&quot;. The list of classes is under the control of the
     database administrator and is not yet available through this
     interface.</pre>
</dd>
<dd>
<pre>
   + name=Class:Landmark:start..end</pre>
</dd>
<dd>
<pre>
     As above, but restricted to the designated range.</pre>
</dd>
<dd>
<p>If you use multiple name options, then this script will generate an
overview image showing the position of each landmark. The alias ``q''
can be used to shorten the length of the URL.</p>
</dd>
</li>
<dt><strong><a name="item_type">type (Alias: t)</a></strong>

<dd>
<p>This argument lists the feature types to display. The value of this
argument is a list of mnemonics separated by spaces (``+'' characters
when URL-escaped). For example:</p>
</dd>
<dd>
<pre>
    &lt;img src=&quot;<a href="http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=mec-3">http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=mec-3</a>;
                 type=tRNA+NG+WABA+CG+ESTB&quot;&gt;</pre>
</dd>
<dd>
<p>Multiple type= arguments will be combined to form a single
space-delimited list. The alias ``t'' can be used to shorten the length
of the URL.</p>
</dd>
</li>
<dt><strong><a name="item_width">width (Alias: w)</a></strong>

<dd>
<p>Width of the desired image, in pixels.</p>
</dd>
</li>
<dt><strong><a name="item_options">options (Alias: o)</a></strong>

<dd>
<p>A space-delimited list (``+'' characters when URL-escaped) of
mnemonic/option pairs describing how features should be
formatted. Options are integers from 0 to 3, where 0=auto, 1=compact,
2=expanded, 3=expanded and labeled. For example, to specify that the
tRNA and NG tracks should always be expanded and labeled, but that the
WABA track should be compact, use:</p>
</dd>
<dd>
<pre>
  options=tRNA+3+NG+3+WABA+1</pre>
</dd>
<dd>
<p>The alias ``o'' can be used to shorten the length of the URL.</p>
</dd>
</li>
<dt><strong><a name="item_abs">abs</a></strong>

<dd>
<p>Use absolute (chromosome) coordinates when displaying the image. An
argument of abs=1 will turn on absolute coordinate addressing. An
argument of abs=0 will use relative coordinate addressing (this is the
default).</p>
</dd>
</li>
<dt><strong><a name="item_add">add (Alias: a)</a></strong>

<dd>
<p>Superimpose one or more additional features on top of the view.
Features are specified as space (``+'') delimited lists in the following
format:</p>
</dd>
<dd>
<pre>
  add=Landmark+Type+Name+start..end,start..end,start..end</pre>
</dd>
<dd>
<p>``Landmark'' is the landmark name, ``Type'' is a descriptive type that
will be printed in the image caption, ``Name'' is a name for the feature
to be printed above it, and start..end is a comma-delimited list of
ranges for discontinuous feature. Names that contain white space must
be quoted, for example ``BLAST hit''. Note that this all has to be
URL-escaped, so an additional feature named ``Your Sequence'', type
``Blast Hit'', that is located on chromosome III in a gapped range
between 20000 and 22000, will be formatted as:</p>
</dd>
<dd>
<pre>
  add=III+%22Blast%20Hit%22+%22Your%20Sequence%22+20000..21000,21550..22000</pre>
</dd>
<dd>
<p>One or both of the type and name can be omitted. If omitted, type will
default to ``Your Features'' and the name will default to ``Feature XX''
where XX is an integer. This allows for a very simple feature line:</p>
</dd>
<dd>
<pre>
  add=III+20000..21000,21550..22000</pre>
</dd>
<dd>
<p>Multiple add= arguments are allowed. The alias ``a'' can be used to
shorten the length of the URL.</p>
</dd>
</li>
<dt><strong><a name="item_style">style</a></strong>

<dd>
<p>The style argument can be used to control the rendering of additional
features added with ``add''. It is a flattened version of the style
configuration sections described in this document For example, if you
have added a ``Blast Hit'' annotation, then you can tell the renderer to
use a red arrow for this glyph in this way:</p>
</dd>
<dd>
<pre>
  style=%22Blast%20Hit%22+glyph=arrow+fgcolor=red</pre>
</dd>
</li>
<dt><strong><a name="item_keystyle">keystyle (Alias: k)</a></strong>

<dd>
<p>Controls the positioning of the track key. One of ``right'', ``left'',
``between'' (default) or ``bottom''</p>
</dd>
</li>
<dt><strong><a name="item_overview">overview</a></strong>

<dd>
<p>Ordinarily the image will show the detail panel if the query region
corresponds to a single region, and the overview panel if multiple
regions match (or if a region that is too large to show
matches). Setting overview=1 will force the overview to be shown in
all cases.</p>
</dd>
</li>
<dt><strong><a name="item_flip">flip (Alias: f)</a></strong>

<dd>
<p>Flip the image left to right. Arguments are 0=don't flip (default),
and 1=flip.</p>
</dd>
</li>
<dt><strong><a name="item_embed">embed</a></strong>

<dd>
<p>Generate image and a corresponding HTML imagemap in a form suitable
for embedding into a frame.</p>
</dd>
</li>
<dt><strong><a name="item_format">format</a></strong>

<dd>
<p>Specify the format for the image file. Either ``GD'' (the default) or
``GD::SVG'' for scaleable vector graphics.</p>
</dd>
</li>
<dt><strong><a name="item_list">list</a></strong>

<dd>
<p>If this argument is present, it will cause the script to dump out
various types of information in plain text form. Currently the two
values for this argument are sources, to dump out the list of data
sources, and types, to dump out the list of configured types. For
list=sources, the script will return a simple text list of the data
source names. For list=types, the script will return a three-column
tab-delimited list giving the track names and feature types
corresponding to the currently-selected data source. The format is as
follows:</p>
</dd>
<dd>
<pre>
  Mnemonic &lt;tab&gt; Full description of feature &lt;tab&gt; [default]</pre>
</dd>
<dd>
<p>The third column contains the word ``default'' if the track will be
shown by default when no type argument is provided.</p>
</dd>
</li>
<dt><strong><a name="item_source_this_argument_specifies_the_database_for_th">source
          This argument specifies the database for the images. The list
          of sources can be found using list=sources.</a></strong>

</dl>
<p>Putting it all together, here's a working (very long) URL:</p>
<p><a href="http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=B0001;add=B0001+pcr+pcr1+20000..333000;add=B0001+%22cool%20knockout%22+kn2+30000..20000,10000..5000;type=add+CG+WTP;style=pcr+glyph=primers;style=%22cool%20knockout%22+glyph=transcript2+bgcolor=orange;abs=1">http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=B0001;add=B0001+pcr+pcr1+20000..333000;add=B0001+%22cool%20knockout%22+kn2+30000..20000,10000..5000;type=add+CG+WTP;style=pcr+glyph=primers;style=%22cool%20knockout%22+glyph=transcript2+bgcolor=orange;abs=1</a></p>
<p>If you wish to associate the image with an imagemap so that clicking
on a feature takes the user to the destination configured in the
gbrowse config file, you may do so by placing the URL in an &lt;iframe&gt;
section and using the embed=1 flag:</p>
<pre>
 &lt;iframe src=&quot;<a href="http://localhost/cgi-bin/gbrowse_img/wormbase?name=B0001;embed=1&quot;width=&quot;100%&quot">http://localhost/cgi-bin/gbrowse_img/wormbase?name=B0001;embed=1&quot;width=&quot;100%&quot</a>; height=&quot;250&quot;&gt;
    &lt;img src=&quot;<a href="http://localhost/cgi-bin/gbrowse_img/wormbase?name=B0001&quot;/&gt">http://localhost/cgi-bin/gbrowse_img/wormbase?name=B0001&quot;/&gt</a>;
 &lt;/iframe&gt;</pre>
<p>Placing an &lt;img&gt; tag inside the &lt;iframe&gt; tag arranges for older
browsers that don't know about iframes to display the static image
instead. You may need to adjust the width and height attributes in
order to avoid browsers placing scrollbars around the frame.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<p>Lincoln Stein <a href="mailto:lstein@cshl.org">lstein@cshl.org</a></p>
<p>Copyright (c) 2002-2004 Cold Spring Harbor Laboratory</p>
<p>This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.</p>
<p>For additional help, see The GMOD and GBrowse pages at <a href="http://gmod.org">http://gmod.org</a>.</p>

</body>

</html>
